{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4eb38ea4-98fa-4da1-9071-2c65d38291ff",
   "metadata": {},
   "source": [
    "# Process plate counts to get fraction infectivities and fit curves\n",
    "This notebook is designed to be run using `snakemake`, and analyzes a plate of sequencing-based neutralization assays.\n",
    "\n",
    "The plots generated by this notebook are interactive, so you can mouseover points for details, use the mouse-scroll to zoom and pan, and use interactive dropdowns at the bottom of the plots.\n",
    "\n",
    "## Setup\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "900fe170-1b4b-4646-b9ee-01d75a76f664",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import sys\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import neutcurve\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import ruamel.yaml as yaml\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc47332-2e29-4605-bf8b-463edb63cac6",
   "metadata": {},
   "source": [
    "Get the variables passed by `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "199f1499-8e9e-4a01-a41c-e7937c94995c",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_csvs = snakemake.input.count_csvs\n",
    "fate_csvs = snakemake.input.fate_csvs\n",
    "viral_library_csv = snakemake.input.viral_library_csv\n",
    "neut_standard_set_csv = snakemake.input.neut_standard_set_csv\n",
    "qc_drops_yaml = snakemake.output.qc_drops\n",
    "frac_infectivity_csv = snakemake.output.frac_infectivity_csv\n",
    "fits_csv = snakemake.output.fits_csv\n",
    "fits_pickle = snakemake.output.fits_pickle\n",
    "samples = snakemake.params.samples\n",
    "plate = snakemake.wildcards.plate\n",
    "plate_params = snakemake.params.plate_params\n",
    "\n",
    "# get thresholds turning lists into tuples as needed\n",
    "manual_drops = {\n",
    "    filter_type: [tuple(w) if isinstance(w, list) else w for w in filter_drops]\n",
    "    for (filter_type, filter_drops) in plate_params[\"manual_drops\"].items()\n",
    "}\n",
    "group = plate_params[\"group\"]\n",
    "qc_thresholds = plate_params[\"qc_thresholds\"]\n",
    "curvefit_params = plate_params[\"curvefit_params\"]\n",
    "curvefit_qc = plate_params[\"curvefit_qc\"]\n",
    "curvefit_qc[\"barcode_serum_replicates_ignore_curvefit_qc\"] = [\n",
    "    tuple(w) for w in curvefit_qc[\"barcode_serum_replicates_ignore_curvefit_qc\"]\n",
    "]\n",
    "\n",
    "print(f\"Processing {plate=}\")\n",
    "\n",
    "samples_df = pd.DataFrame(plate_params[\"samples\"])\n",
    "print(f\"\\nPlate has {len(samples)} samples (wells)\")\n",
    "assert all(\n",
    "    (len(samples_df) == samples_df[c].nunique())\n",
    "    for c in [\"well\", \"sample\", \"sample_noplate\"]\n",
    ")\n",
    "assert len(samples_df) == len(\n",
    "    samples_df.groupby([\"serum_replicate\", \"dilution_factor\"])\n",
    ")\n",
    "assert len(samples) == len(count_csvs) == len(fate_csvs) == len(samples_df)\n",
    "\n",
    "for d, key, title in [\n",
    "    (manual_drops, \"manual_drops\", \"Data manually specified to drop:\"),\n",
    "    (qc_thresholds, \"qc_thresholds\", \"QC thresholds applied to data:\"),\n",
    "    (curvefit_params, \"curvefit_params\", \"Curve-fitting parameters:\"),\n",
    "    (curvefit_qc, \"curvefit_qc\", \"Curve-fitting QC:\"),\n",
    "]:\n",
    "    print(f\"\\n{title}\")\n",
    "    yaml.YAML(typ=\"rt\").dump({key: d}, stream=sys.stdout)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbf4bda2-e16f-4fc3-adbb-b5d3a7bfe2e5",
   "metadata": {},
   "source": [
    "Set up dictionary to keep track of wells, barcodes, well-barcodes, and serum-replicates that are dropped:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aab62a4-f32f-49cc-bb4e-0e96d5a2d1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_drops = {\n",
    "    \"wells\": {},\n",
    "    \"barcodes\": {},\n",
    "    \"barcode_wells\": {},\n",
    "    \"barcode_serum_replicates\": {},\n",
    "    \"serum_replicates\": {},\n",
    "}\n",
    "\n",
    "assert set(manual_drops).issubset(\n",
    "    qc_drops\n",
    "), f\"{manual_drops.keys()=}, {qc_drops.keys()}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc6f5cd-de3d-4824-a421-7b460090d72e",
   "metadata": {},
   "source": [
    "## Statistics on barcode-parsing for each sample\n",
    "Make interactive chart of the \"fates\" of the sequencing reads parsed for each sample on the plate.\n",
    "\n",
    "If most sequencing reads are not \"valid barcodes\", this could potentially indicate some problem in the sequencing or barcode set you are parsing.\n",
    "\n",
    "Potential fates are:\n",
    " - *valid barcode*: barcode that matches a known virus or neutralization standard, we hope most reads are this.\n",
    " - *invalid barcode*: a barcode with proper flanking sequences, but does not match a known virus or neutralization standard. If you  have a lot of reads of this type, it is probably a good idea to look at the invalid barcode CSVs (in the `./results/barcode_invalid/` subdirectory created by the pipeline) to see what these invalid barcodes are.\n",
    " - *unparseable barcode*: could not parse a barcode from this read as there was not a sequence of the correct length with the appropriate flanking sequence.\n",
    " - *invalid outer flank*: if using an outer upstream or downstream region (`upstream2` or `downstream2` for the [illuminabarcodeparser](https://jbloomlab.github.io/dms_variants/dms_variants.illuminabarcodeparser.html)), reads that are otherwise valid except for this outer flank. Typically you would be using `upstream2` if you have a plate index embedded in your primer, and reads with this classification correspond to a different index than the one for this plate.\n",
    " - *low quality barcode*: low-quality or `N` nucleotides in barcode, could indicate problem with sequencing.\n",
    " - *failed chastity filter*: reads that failed the Illumina chastity filter, if these are reported in the FASTQ (they may not be).\n",
    "\n",
    "Also, if the number of reads per sample is very uneven, that could indicate that you did not do a good job of balancing the different samples in the Illumina sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa103c7-3240-4dac-8532-f7298a5fe009",
   "metadata": {},
   "outputs": [],
   "source": [
    "fates = (\n",
    "    pd.concat([pd.read_csv(f).assign(sample=s) for f, s in zip(fate_csvs, samples)])\n",
    "    .merge(samples_df, validate=\"many_to_one\", on=\"sample\")\n",
    "    .assign(\n",
    "        fate_counts=lambda x: x.groupby(\"fate\")[\"count\"].transform(\"sum\"),\n",
    "        sample_well=lambda x: x[\"sample_noplate\"] + \" (\" + x[\"well\"] + \")\",\n",
    "    )\n",
    "    .query(\"fate_counts > 0\")[  # only keep fates with at least one count\n",
    "        [\"fate\", \"count\", \"well\", \"serum_replicate\", \"sample_well\", \"dilution_factor\"]\n",
    "    ]\n",
    ")\n",
    "\n",
    "assert len(fates) == len(fates.drop_duplicates())\n",
    "\n",
    "serum_replicates = sorted(fates[\"serum_replicate\"].unique())\n",
    "sample_wells = list(\n",
    "    fates.sort_values([\"serum_replicate\", \"dilution_factor\"])[\"sample_well\"]\n",
    ")\n",
    "\n",
    "\n",
    "serum_selection = alt.selection_point(\n",
    "    fields=[\"serum_replicate\"],\n",
    "    bind=alt.binding_select(\n",
    "        options=[None] + serum_replicates,\n",
    "        labels=[\"all\"] + serum_replicates,\n",
    "        name=\"serum\",\n",
    "    ),\n",
    ")\n",
    "\n",
    "fates_chart = (\n",
    "    alt.Chart(fates)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\"count\", scale=alt.Scale(nice=False, padding=3)),\n",
    "        alt.Y(\n",
    "            \"sample_well\",\n",
    "            title=None,\n",
    "            sort=sample_wells,\n",
    "        ),\n",
    "        alt.Color(\"fate\", sort=sorted(fates[\"fate\"].unique(), reverse=True)),\n",
    "        alt.Order(\"fate\", sort=\"descending\"),\n",
    "        tooltip=fates.columns.tolist(),\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=200,\n",
    "        title=f\"Barcode parsing for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "fates_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a810fb9a-189d-48e5-b866-6ef0f2e0dd5f",
   "metadata": {},
   "source": [
    "## Read barcode counts and apply manually specified drops\n",
    "Read the counts per barcode:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccea55e6-d280-4e8e-998a-1a86930c6529",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get barcode counts\n",
    "counts = (\n",
    "    pd.concat([pd.read_csv(c).assign(sample=s) for c, s in zip(count_csvs, samples)])\n",
    "    .merge(samples_df, validate=\"many_to_one\", on=\"sample\")\n",
    "    .drop(columns=[\"replicate\", \"plate\", \"fastq\"])\n",
    "    .assign(sample_well=lambda x: x[\"sample_noplate\"] + \" (\" + x[\"well\"] + \")\")\n",
    ")\n",
    "\n",
    "# classify barcodes as viral or neut standard\n",
    "barcode_class = pd.concat(\n",
    "    [\n",
    "        pd.read_csv(viral_library_csv)[[\"barcode\", \"strain\"]].assign(\n",
    "            neut_standard=False,\n",
    "        ),\n",
    "        pd.read_csv(neut_standard_set_csv)[[\"barcode\"]].assign(\n",
    "            neut_standard=True,\n",
    "            strain=pd.NA,\n",
    "        ),\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "# merge counts and classification of barcodes\n",
    "assert set(counts[\"barcode\"]) == set(barcode_class[\"barcode\"])\n",
    "counts = counts.merge(barcode_class, on=\"barcode\", validate=\"many_to_one\")\n",
    "assert set(sample_wells) == set(counts[\"sample_well\"])\n",
    "assert set(serum_replicates) == set(counts[\"serum_replicate\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1326e88-a654-4a21-b9e5-85f5bf93847f",
   "metadata": {},
   "source": [
    "Apply any manually specified data drops:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9199cc2e-b6c8-4dbc-8d2f-ec350da71c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for filter_type, filter_drops in manual_drops.items():\n",
    "    print(f\"\\nDropping {len(filter_drops)} {filter_type} specified in manual_drops\")\n",
    "    assert filter_type in qc_drops\n",
    "    qc_drops[filter_type].update(\n",
    "        {w: \"manual_drop\" for w in filter_drops if not isinstance(w, list)}\n",
    "    )\n",
    "    if filter_type == \"barcode_wells\":\n",
    "        counts = counts[\n",
    "            ~counts.assign(\n",
    "                barcode_well=lambda x: x.apply(\n",
    "                    lambda r: (r[\"barcode\"], r[\"well\"]), axis=1\n",
    "                )\n",
    "            )[\"barcode_well\"].isin(qc_drops[filter_type])\n",
    "        ]\n",
    "    elif filter_type == \"barcode_serum_replicates\":\n",
    "        counts = counts[\n",
    "            ~counts.assign(\n",
    "                barcode_serum_replicate=lambda x: x.apply(\n",
    "                    lambda r: (r[\"barcode\"], r[\"serum_replicate\"]), axis=1\n",
    "                )\n",
    "            )[\"barcode_serum_replicate\"].isin(qc_drops[filter_type])\n",
    "        ]\n",
    "    elif filter_type == \"wells\":\n",
    "        counts = counts[~counts[\"well\"].isin(qc_drops[filter_type])]\n",
    "    elif filter_type == \"barcodes\":\n",
    "        counts = counts[~counts[\"barcode\"].isin(qc_drops[filter_type])]\n",
    "    elif filter_type == \"serum_replicates\":\n",
    "        counts = counts[~counts[\"serum_replicate\"].isin(qc_drops[filter_type])]\n",
    "    elif filter_type == \"barcode_serum_replicates\":\n",
    "        counts = counts[~counts[\"barcode_serum_replicate\"].isin(qc_drops[filter_type])]\n",
    "    else:\n",
    "        assert filter_type in set(counts.columns)\n",
    "        counts = counts[~counts[filter_type].isin(qc_drops[filter_type])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faedb6a2-ebcd-4d56-8ce9-dddfc2a0569d",
   "metadata": {},
   "source": [
    "## Average counts per barcode in each well"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ddb2544-a1a3-460f-9d81-fa061b152071",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-05T19:26:10.449100Z",
     "iopub.status.busy": "2023-12-05T19:26:10.448471Z",
     "iopub.status.idle": "2023-12-05T19:26:10.468023Z",
     "shell.execute_reply": "2023-12-05T19:26:10.467117Z",
     "shell.execute_reply.started": "2023-12-05T19:26:10.449041Z"
    }
   },
   "source": [
    "Plot average counts per barcode.\n",
    "If a sample has inadequate barcode counts, it may not have good enough statistics for accurate analysis, and a QC-threshold is applied:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68da99a3-e1e4-4b03-a7c2-79406e6d7170",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_barcode_counts = (\n",
    "    counts.groupby(\n",
    "        [\"well\", \"serum_replicate\", \"sample_well\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(avg_count=pd.NamedAgg(\"count\", \"mean\"))\n",
    "    .assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"avg_count\"] < qc_thresholds[\"avg_barcode_counts_per_well\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "avg_barcode_counts_chart = (\n",
    "    alt.Chart(avg_barcode_counts)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"avg_count\",\n",
    "            title=\"average barcode counts per well\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\"sample_well\", sort=sample_wells),\n",
    "        alt.Color(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['avg_barcode_counts_per_well']=}\",\n",
    "            legend=alt.Legend(titleLimit=500),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if avg_barcode_counts[c].dtype == float else c\n",
    "            for c in avg_barcode_counts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Average barcode counts per well for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "display(avg_barcode_counts_chart)\n",
    "\n",
    "# drop wells failing QC\n",
    "avg_barcode_counts_per_well_drops = list(avg_barcode_counts.query(\"fails_qc\")[\"well\"])\n",
    "print(\n",
    "    f\"\\nDropping {len(avg_barcode_counts_per_well_drops)} wells for failing \"\n",
    "    f\"{qc_thresholds['avg_barcode_counts_per_well']=}: \"\n",
    "    + str(avg_barcode_counts_per_well_drops)\n",
    ")\n",
    "qc_drops[\"wells\"].update(\n",
    "    {w: \"avg_barcode_counts_per_well\" for w in avg_barcode_counts_per_well_drops}\n",
    ")\n",
    "counts = counts[~counts[\"well\"].isin(qc_drops[\"wells\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589a755d-b46f-46ea-a84c-2dd784fa8672",
   "metadata": {},
   "source": [
    "## Fraction of counts from neutralization standard\n",
    "Determine the fraction of counts from the neutralization standard in each sample, and make sure this fraction passess the QC threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a71183-4670-40c1-88c1-237bb536803e",
   "metadata": {},
   "outputs": [],
   "source": [
    "neut_standard_fracs = (\n",
    "    counts.assign(\n",
    "        neut_standard_count=lambda x: x[\"count\"] * x[\"neut_standard\"].astype(int)\n",
    "    )\n",
    "    .groupby(\n",
    "        [\"well\", \"serum_replicate\", \"sample_well\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(\n",
    "        total_count=pd.NamedAgg(\"count\", \"sum\"),\n",
    "        neut_standard_count=pd.NamedAgg(\"neut_standard_count\", \"sum\"),\n",
    "    )\n",
    "    .assign(\n",
    "        neut_standard_frac=lambda x: x[\"neut_standard_count\"] / x[\"total_count\"],\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"neut_standard_frac\"] < qc_thresholds[\"min_neut_standard_frac_per_well\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "neut_standard_fracs_chart = (\n",
    "    alt.Chart(neut_standard_fracs)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"neut_standard_frac\",\n",
    "            title=\"frac counts from neutralization standard per well\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\"sample_well\", sort=sample_wells),\n",
    "        alt.Color(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['min_neut_standard_frac_per_well']=}\",\n",
    "            legend=alt.Legend(titleLimit=500),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if neut_standard_fracs[c].dtype == float else c\n",
    "            for c in neut_standard_fracs.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Neutralization-standard fracs per well for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(neut_standard_fracs_chart)\n",
    "\n",
    "# drop wells failing QC\n",
    "min_neut_standard_frac_per_well_drops = list(\n",
    "    neut_standard_fracs.query(\"fails_qc\")[\"well\"]\n",
    ")\n",
    "print(\n",
    "    f\"\\nDropping {len(min_neut_standard_frac_per_well_drops)} wells for failing \"\n",
    "    f\"{qc_thresholds['min_neut_standard_frac_per_well']=}: \"\n",
    "    + str(min_neut_standard_frac_per_well_drops)\n",
    ")\n",
    "qc_drops[\"wells\"].update(\n",
    "    {\n",
    "        w: \"min_neut_standard_frac_per_well\"\n",
    "        for w in min_neut_standard_frac_per_well_drops\n",
    "    }\n",
    ")\n",
    "counts = counts[~counts[\"well\"].isin(qc_drops[\"wells\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e990a5d-075e-4e81-971c-14483f1dd7ea",
   "metadata": {},
   "source": [
    "## Consistency and minimum fractions for barcodes\n",
    "We examine the fraction of counts attributable to each barcode. We do this splitting the data two ways:\n",
    "\n",
    " 1. Looking at all viral (but not neut-standard) barcodes only for the no-serum samples (wells).\n",
    "\n",
    " 2. Looking at just the neut-standard barcodes for all samples (wells).\n",
    "\n",
    "The reasons is that if the experiment is set up perfectly, these fractions should be the same across all samples for each barcode.\n",
    "(We do not expect viral barcodes to have consistent fractions across no-serum samples as they will be neutralized differently depending on strain).\n",
    "\n",
    "We plot these fractions in interactive plots (you can mouseover points and zoom) so you can identify barcodes that fail the expected consistency QC thresholds.\n",
    "\n",
    "We also make sure the barcodes meet specified QC minimum thresholds for all samples, and flag any that do not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fecec428-ad2f-4877-bcd4-a399fc7b7d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_selection = alt.selection_point(fields=[\"barcode\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "# look at all samples for neut standard barcodes, or no-serum samples for all barcodes\n",
    "for is_neut_standard, df in counts.groupby(\"neut_standard\"):\n",
    "    if is_neut_standard:\n",
    "        print(\n",
    "            f\"\\n\\n{'=' * 89}\\nAnalyzing neut-standard barcodes from all samples (wells)\"\n",
    "        )\n",
    "        qc_name = \"per_neut_standard_barcode_filters\"\n",
    "    else:\n",
    "        print(f\"\\n\\n{'=' * 89}\\nAnalyzing all barcodes from no-serum samples (wells)\")\n",
    "        qc_name = \"no_serum_per_viral_barcode_filters\"\n",
    "        df = df.query(\"serum == 'none'\")\n",
    "\n",
    "    df = df.assign(\n",
    "        sample_counts=lambda x: x.groupby(\"sample\")[\"count\"].transform(\"sum\"),\n",
    "        count_frac=lambda x: x[\"count\"] / x[\"sample_counts\"],\n",
    "        median_count_frac=lambda x: x.groupby(\"barcode\")[\"count_frac\"].transform(\n",
    "            \"median\"\n",
    "        ),\n",
    "        fold_change_from_median=lambda x: numpy.where(\n",
    "            x[\"count_frac\"] > x[\"median_count_frac\"],\n",
    "            x[\"count_frac\"] / x[\"median_count_frac\"],\n",
    "            x[\"median_count_frac\"] / x[\"count_frac\"],\n",
    "        ),\n",
    "    )[\n",
    "        [\n",
    "            \"barcode\",\n",
    "            \"count\",\n",
    "            \"well\",\n",
    "            \"sample_well\",\n",
    "            \"count_frac\",\n",
    "            \"median_count_frac\",\n",
    "            \"fold_change_from_median\",\n",
    "        ]\n",
    "        + ([] if is_neut_standard else [\"strain\"])\n",
    "    ]\n",
    "\n",
    "    # barcode fails QC if fails in sufficient wells\n",
    "    qc = qc_thresholds[qc_name]\n",
    "    print(f\"Apply QC {qc_name}: {qc}\\n\")\n",
    "    fails_qc = (\n",
    "        df.assign(\n",
    "            fails_qc=lambda x: ~(\n",
    "                (x[\"count_frac\"] >= qc[\"min_frac\"])\n",
    "                & (x[\"fold_change_from_median\"] <= qc[\"max_fold_change\"])\n",
    "            ),\n",
    "        )\n",
    "        .groupby(\"barcode\", as_index=False)\n",
    "        .aggregate(n_wells_fail_qc=pd.NamedAgg(\"fails_qc\", \"sum\"))\n",
    "        .assign(fails_qc=lambda x: x[\"n_wells_fail_qc\"] >= qc[\"max_wells\"])[\n",
    "            [\"barcode\", \"fails_qc\"]\n",
    "        ]\n",
    "    )\n",
    "    df = df.merge(fails_qc, on=\"barcode\", validate=\"many_to_one\")\n",
    "\n",
    "    # make chart\n",
    "    evenness_chart = (\n",
    "        alt.Chart(df)\n",
    "        .add_params(barcode_selection)\n",
    "        .encode(\n",
    "            alt.X(\n",
    "                \"count_frac\",\n",
    "                title=(\n",
    "                    \"barcode's fraction of neut standard counts\"\n",
    "                    if is_neut_standard\n",
    "                    else \"barcode's fraction of non-neut standard counts\"\n",
    "                ),\n",
    "                scale=alt.Scale(nice=False, padding=5),\n",
    "            ),\n",
    "            alt.Y(\"sample_well\", sort=sample_wells),\n",
    "            alt.Fill(\n",
    "                \"fails_qc\",\n",
    "                title=f\"fails {qc_name}\",\n",
    "                legend=alt.Legend(titleLimit=500),\n",
    "            ),\n",
    "            strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "            size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".2g\") if df[c].dtype == float else c\n",
    "                for c in df.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_circle(fillOpacity=0.45, stroke=\"black\", strokeOpacity=1)\n",
    "        .properties(\n",
    "            height=alt.Step(10),\n",
    "            width=300,\n",
    "            title=alt.TitleParams(\n",
    "                (\n",
    "                    f\"{plate} all samples, neut-standard barcodes\"\n",
    "                    if is_neut_standard\n",
    "                    else f\"{plate} no-serum samples, all barcodes\"\n",
    "                ),\n",
    "                subtitle=\"x-axis is zoomable (use mouse scroll/pan)\",\n",
    "            ),\n",
    "        )\n",
    "        .configure_axis(grid=False)\n",
    "        .configure_legend(titleLimit=1000)\n",
    "        .interactive()\n",
    "    )\n",
    "\n",
    "    display(evenness_chart)\n",
    "\n",
    "    # drop barcodes failing QC\n",
    "    barcode_drops = list(fails_qc.query(\"fails_qc\")[\"barcode\"])\n",
    "    print(\n",
    "        f\"\\nDropping {len(barcode_drops)} barcodes for failing {qc=}: {barcode_drops}\"\n",
    "    )\n",
    "    qc_drops[\"barcodes\"].update(\n",
    "        {bc: \"min_neut_standard_frac_per_well\" for bc in barcode_drops}\n",
    "    )\n",
    "    counts = counts[~counts[\"barcode\"].isin(qc_drops[\"barcodes\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743e4e41-aef9-4980-8b59-55d7b0a686fd",
   "metadata": {},
   "source": [
    "## Compute fraction infectivity\n",
    "\n",
    "The fraction infectivity for viral barcode $v_b$ in sample $s$ is computed as:\n",
    "$$\n",
    "F_{v_b,s} = \\frac{c_{v_b,s} / \\left(\\sum_{n_b} c_{n_b,s}\\right)}{{\\rm median}_{s_0}\\left[ c_{v_b,s_0} / \\left(\\sum_{n_b} c_{n_b,s_0}\\right)\\right]}\n",
    "$$\n",
    "where\n",
    " - $c_{v_b,s}$ is the counts of viral barcode $v_b$ in sample $s$.\n",
    " - $\\sum_{n_b} c_{n_b,s}$ is the sum of the counts for all neutralization standard barcodes $n_b$ for sample $s$.\n",
    " - $c_{v_b,s_0}$ is the counts of viral barcode $v_b$ in no-serum sample $s_0$.\n",
    " - $\\sum_{n_b} c_{n_b,s_0}$ is the sum of the counts for all neutralization standard barcodes $n_b$ for no-serum sample $s_0$.\n",
    " - ${\\rm median}_{s_0}\\left[ c_{v_b,s_0} / \\left(\\sum_{n_b} c_{n_b,s_0}\\right)\\right]$ is the median taken across all no-serum samples of the counts of viral barcode $v_b$ versus the total counts for all neutralization standard barcodes.\n",
    "\n",
    "First, compute the total neutralization-standard counts for each sample (well).\n",
    "Plot these, and drop any wells that do not meet the QC threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ef1afa4-5772-4596-a68d-b98e080ba2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "neut_standard_counts = (\n",
    "    counts.query(\"neut_standard\")\n",
    "    .groupby(\n",
    "        [\"well\", \"serum_replicate\", \"sample_well\", \"dilution_factor\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(neut_standard_count=pd.NamedAgg(\"count\", \"sum\"))\n",
    "    .assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"neut_standard_count\"] < qc_thresholds[\"min_neut_standard_count_per_well\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "neut_standard_counts_chart = (\n",
    "    alt.Chart(neut_standard_counts)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"neut_standard_count\",\n",
    "            title=\"counts from neutralization standard\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\"sample_well\", sort=sample_wells),\n",
    "        alt.Color(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['min_neut_standard_count_per_well']=}\",\n",
    "            legend=alt.Legend(titleLimit=500),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            (\n",
    "                alt.Tooltip(c, format=\".3g\")\n",
    "                if neut_standard_counts[c].dtype == float\n",
    "                else c\n",
    "            )\n",
    "            for c in neut_standard_counts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Neutralization-standard counts for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(neut_standard_counts_chart)\n",
    "\n",
    "# drop wells failing QC\n",
    "min_neut_standard_count_per_well_drops = list(\n",
    "    neut_standard_counts.query(\"fails_qc\")[\"well\"]\n",
    ")\n",
    "print(\n",
    "    f\"\\nDropping {len(min_neut_standard_count_per_well_drops)} wells for failing \"\n",
    "    f\"{qc_thresholds['min_neut_standard_count_per_well']=}: \"\n",
    "    + str(min_neut_standard_count_per_well_drops)\n",
    ")\n",
    "qc_drops[\"wells\"].update(\n",
    "    {\n",
    "        w: \"min_neut_standard_count_per_well\"\n",
    "        for w in min_neut_standard_count_per_well_drops\n",
    "    }\n",
    ")\n",
    "neut_standard_counts = neut_standard_counts[\n",
    "    ~neut_standard_counts[\"well\"].isin(qc_drops[\"wells\"])\n",
    "]\n",
    "counts = counts[~counts[\"well\"].isin(qc_drops[\"wells\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9baafb57-e601-46d4-bed4-4bd5eb61552b",
   "metadata": {},
   "source": [
    "Compute and plot the no-serum sample viral barcode counts and check if they pass the QC filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d060ad12-b159-49e0-8ff4-94d175da5bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_serum_counts = (\n",
    "    counts.query(\"serum == 'none'\")\n",
    "    .query(\"not neut_standard\")\n",
    "    .merge(neut_standard_counts, validate=\"many_to_one\")[\n",
    "        [\"barcode\", \"strain\", \"well\", \"sample_well\", \"count\", \"neut_standard_count\"]\n",
    "    ]\n",
    "    .assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"count\"] <= qc_thresholds[\"min_no_serum_count_per_viral_barcode_well\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "strains = sorted(no_serum_counts[\"strain\"].unique())\n",
    "strain_selection_dropdown = alt.selection_point(\n",
    "    fields=[\"strain\"],\n",
    "    bind=alt.binding_select(\n",
    "        options=[None] + strains,\n",
    "        labels=[\"all\"] + strains,\n",
    "        name=\"virus strain\",\n",
    "    ),\n",
    ")\n",
    "\n",
    "# make chart\n",
    "no_serum_counts_chart = (\n",
    "    alt.Chart(no_serum_counts)\n",
    "    .add_params(barcode_selection, strain_selection_dropdown)\n",
    "    .transform_filter(strain_selection_dropdown)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"count\", title=\"viral barcode count\", scale=alt.Scale(nice=False, padding=5)\n",
    "        ),\n",
    "        alt.Y(\"sample_well\", sort=sample_wells),\n",
    "        alt.Fill(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['min_no_serum_count_per_viral_barcode_well']=}\",\n",
    "            legend=alt.Legend(titleLimit=500),\n",
    "        ),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "        size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "        tooltip=no_serum_counts.columns.tolist(),\n",
    "    )\n",
    "    .mark_circle(fillOpacity=0.6, stroke=\"black\", strokeOpacity=1)\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=400,\n",
    "        title=f\"{plate} viral barcode counts in no-serum samples\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    "    .interactive()\n",
    ")\n",
    "\n",
    "display(no_serum_counts_chart)\n",
    "\n",
    "# drop barcode / wells failing QC\n",
    "min_no_serum_count_per_viral_barcode_well_drops = list(\n",
    "    no_serum_counts.query(\"fails_qc\")[[\"barcode\", \"well\"]].itertuples(\n",
    "        index=False, name=None\n",
    "    )\n",
    ")\n",
    "print(\n",
    "    f\"\\nDropping {len(min_no_serum_count_per_viral_barcode_well_drops)} barcode-wells for failing \"\n",
    "    f\"{qc_thresholds['min_no_serum_count_per_viral_barcode_well']=}: \"\n",
    "    + str(min_no_serum_count_per_viral_barcode_well_drops)\n",
    ")\n",
    "qc_drops[\"barcode_wells\"].update(\n",
    "    {\n",
    "        w: \"min_no_serum_count_per_viral_barcode_well\"\n",
    "        for w in min_no_serum_count_per_viral_barcode_well_drops\n",
    "    }\n",
    ")\n",
    "no_serum_counts = no_serum_counts[\n",
    "    ~no_serum_counts.assign(\n",
    "        barcode_well=lambda x: x.apply(lambda r: (r[\"barcode\"], r[\"well\"]), axis=1)\n",
    "    )[\"barcode_well\"].isin(qc_drops[\"barcode_wells\"])\n",
    "]\n",
    "counts = counts[\n",
    "    ~counts.assign(\n",
    "        barcode_well=lambda x: x.apply(lambda r: (r[\"barcode\"], r[\"well\"]), axis=1)\n",
    "    )[\"barcode_well\"].isin(qc_drops[\"barcode_wells\"])\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdffe0e7-9864-4c4f-851d-5604106ff63a",
   "metadata": {},
   "source": [
    "Compute and plot the median ratio of viral barcode count to neut standard counts across no-serum samples.\n",
    "If library composition is equal, all of these values should be similar:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b15a0c-9d54-48f4-86e3-f38b2015a5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "median_no_serum_ratio = (\n",
    "    no_serum_counts.assign(ratio=lambda x: x[\"count\"] / x[\"neut_standard_count\"])\n",
    "    .groupby([\"barcode\", \"strain\"], as_index=False)\n",
    "    .aggregate(median_no_serum_ratio=pd.NamedAgg(\"ratio\", \"median\"))\n",
    ")\n",
    "\n",
    "strain_selection = alt.selection_point(fields=[\"strain\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "median_no_serum_ratio_chart = (\n",
    "    alt.Chart(median_no_serum_ratio)\n",
    "    .add_params(strain_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"median_no_serum_ratio\",\n",
    "            title=\"median ratio of counts\",\n",
    "            scale=alt.Scale(nice=False, padding=5),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"barcode\",\n",
    "            sort=alt.SortField(\"median_no_serum_ratio\", order=\"descending\"),\n",
    "            axis=alt.Axis(labelFontSize=5),\n",
    "        ),\n",
    "        color=alt.condition(strain_selection, alt.value(\"orange\"), alt.value(\"gray\")),\n",
    "        tooltip=[\n",
    "            (\n",
    "                alt.Tooltip(c, format=\".3g\")\n",
    "                if median_no_serum_ratio[c].dtype == float\n",
    "                else c\n",
    "            )\n",
    "            for c in median_no_serum_ratio.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(5),\n",
    "        width=250,\n",
    "        title=f\"{plate} no-serum median ratio viral barcode to neut-standard barcode\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(median_no_serum_ratio_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb39004a-3684-4501-afd0-da259c85b131",
   "metadata": {},
   "source": [
    "Compute the actual fraction infectivities.\n",
    "We compute both the raw fraction infectivities and the ones with the ceiling applied:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d787193c-3c50-427d-a0ec-165b446d8e73",
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_infectivity = (\n",
    "    counts.query(\"not neut_standard\")\n",
    "    .query(\"serum != 'none'\")\n",
    "    .merge(median_no_serum_ratio, validate=\"many_to_one\")\n",
    "    .merge(neut_standard_counts, validate=\"many_to_one\")\n",
    "    .assign(\n",
    "        frac_infectivity_raw=lambda x: (\n",
    "            (x[\"count\"] / x[\"neut_standard_count\"]) / x[\"median_no_serum_ratio\"]\n",
    "        ),\n",
    "        frac_infectivity_ceiling=lambda x: x[\"frac_infectivity_raw\"].clip(\n",
    "            upper=curvefit_params[\"frac_infectivity_ceiling\"]\n",
    "        ),\n",
    "        concentration=lambda x: 1 / x[\"dilution_factor\"],\n",
    "        plate_barcode=lambda x: x[\"plate_replicate\"] + \"-\" + x[\"barcode\"],\n",
    "    )[\n",
    "        [\n",
    "            \"barcode\",\n",
    "            \"plate_barcode\",\n",
    "            \"well\",\n",
    "            \"strain\",\n",
    "            \"serum\",\n",
    "            \"serum_replicate\",\n",
    "            \"dilution_factor\",\n",
    "            \"concentration\",\n",
    "            \"frac_infectivity_raw\",\n",
    "            \"frac_infectivity_ceiling\",\n",
    "        ]\n",
    "    ]\n",
    ")\n",
    "\n",
    "assert len(\n",
    "    frac_infectivity.groupby([\"serum\", \"plate_barcode\", \"dilution_factor\"])\n",
    ") == len(frac_infectivity)\n",
    "assert frac_infectivity[\"dilution_factor\"].notnull().all()\n",
    "assert frac_infectivity[\"frac_infectivity_raw\"].notnull().all()\n",
    "assert frac_infectivity[\"frac_infectivity_ceiling\"].notnull().all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "465014e0-aced-4fb2-a503-620c1636c804",
   "metadata": {},
   "source": [
    "Plot the fraction infectivities, both the raw values and with the ceiling applied:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982d871b-c140-4295-afe0-ea138bdafbaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_infectivity_chart_df = (\n",
    "    frac_infectivity.assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"frac_infectivity_raw\"]\n",
    "            > qc_thresholds[\"max_frac_infectivity_per_viral_barcode_well\"]\n",
    "        ),\n",
    "    )\n",
    "    .melt(\n",
    "        id_vars=[\n",
    "            \"barcode\",\n",
    "            \"strain\",\n",
    "            \"well\",\n",
    "            \"serum_replicate\",\n",
    "            \"dilution_factor\",\n",
    "            \"fails_qc\",\n",
    "        ],\n",
    "        value_vars=[\"frac_infectivity_raw\", \"frac_infectivity_ceiling\"],\n",
    "        var_name=\"ceiling_applied\",\n",
    "        value_name=\"frac_infectivity\",\n",
    "    )\n",
    "    .assign(\n",
    "        ceiling_applied=lambda x: x[\"ceiling_applied\"].map(\n",
    "            {\n",
    "                \"frac_infectivity_raw\": \"raw fraction infectivity\",\n",
    "                \"frac_infectivity_ceiling\": f\"fraction infectivity with ceiling at {curvefit_params['frac_infectivity_ceiling']}\",\n",
    "            }\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "frac_infectivity_chart = (\n",
    "    alt.Chart(frac_infectivity_chart_df)\n",
    "    .add_params(strain_selection_dropdown, barcode_selection)\n",
    "    .transform_filter(strain_selection_dropdown)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"dilution_factor\",\n",
    "            title=\"dilution factor\",\n",
    "            scale=alt.Scale(nice=False, padding=5, type=\"log\"),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"frac_infectivity\",\n",
    "            title=\"fraction infectivity\",\n",
    "            scale=alt.Scale(nice=False, padding=5),\n",
    "        ),\n",
    "        alt.Column(\n",
    "            \"ceiling_applied\",\n",
    "            sort=\"descending\",\n",
    "            title=None,\n",
    "            header=alt.Header(labelFontSize=13, labelFontStyle=\"bold\", labelPadding=2),\n",
    "        ),\n",
    "        alt.Row(\n",
    "            \"serum_replicate\",\n",
    "            title=None,\n",
    "            spacing=3,\n",
    "            header=alt.Header(labelFontSize=13, labelFontStyle=\"bold\"),\n",
    "        ),\n",
    "        alt.Detail(\"barcode\"),\n",
    "        alt.Shape(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['max_frac_infectivity_per_viral_barcode_well']=}\",\n",
    "            legend=alt.Legend(titleLimit=500, orient=\"bottom\"),\n",
    "        ),\n",
    "        color=alt.condition(\n",
    "            barcode_selection, alt.value(\"black\"), alt.value(\"MediumBlue\")\n",
    "        ),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(3), alt.value(1)),\n",
    "        opacity=alt.condition(barcode_selection, alt.value(1), alt.value(0.25)),\n",
    "        tooltip=[\n",
    "            (\n",
    "                alt.Tooltip(c, format=\".3g\")\n",
    "                if frac_infectivity_chart_df[c].dtype == float\n",
    "                else c\n",
    "            )\n",
    "            for c in frac_infectivity_chart_df.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_line(point=True)\n",
    "    .properties(\n",
    "        height=150,\n",
    "        width=250,\n",
    "        title=f\"Fraction infectivities for {plate}\",\n",
    "    )\n",
    "    .interactive(bind_x=False)\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    "    .configure_point(size=50)\n",
    "    .resolve_scale(x=\"independent\", y=\"independent\")\n",
    ")\n",
    "\n",
    "display(frac_infectivity_chart)\n",
    "\n",
    "# drop barcode / wells failing QC\n",
    "max_frac_infectivity_per_viral_barcode_well_drops = list(\n",
    "    frac_infectivity_chart_df.query(\"fails_qc\")[[\"barcode\", \"well\"]]\n",
    "    .drop_duplicates()\n",
    "    .itertuples(index=False, name=None)\n",
    ")\n",
    "print(\n",
    "    f\"\\nDropping {len(max_frac_infectivity_per_viral_barcode_well_drops)} barcode-wells for failing \"\n",
    "    f\"{qc_thresholds['max_frac_infectivity_per_viral_barcode_well']=}: \"\n",
    "    + str(max_frac_infectivity_per_viral_barcode_well_drops)\n",
    ")\n",
    "qc_drops[\"barcode_wells\"].update(\n",
    "    {\n",
    "        w: \"max_frac_infectivity_per_viral_barcode_well\"\n",
    "        for w in max_frac_infectivity_per_viral_barcode_well_drops\n",
    "    }\n",
    ")\n",
    "frac_infectivity = frac_infectivity[\n",
    "    ~frac_infectivity.assign(\n",
    "        barcode_well=lambda x: x.apply(lambda r: (r[\"barcode\"], r[\"well\"]), axis=1)\n",
    "    )[\"barcode_well\"].isin(qc_drops[\"barcode_wells\"])\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdb4eea-9822-4f27-89e6-6f11033bcb7f",
   "metadata": {},
   "source": [
    "Check how many dilutions we have per barcode / serum-replicate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e8998c6-b2de-43b8-9091-b65423257bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_dilutions = (\n",
    "    frac_infectivity.groupby([\"serum_replicate\", \"strain\", \"barcode\"], as_index=False)\n",
    "    .aggregate(**{\"number of dilutions\": pd.NamedAgg(\"dilution_factor\", \"nunique\")})\n",
    "    .assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"number of dilutions\"]\n",
    "            < qc_thresholds[\"min_dilutions_per_barcode_serum_replicate\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "n_dilutions_chart = (\n",
    "    alt.Chart(n_dilutions)\n",
    "    .add_params(barcode_selection)\n",
    "    .encode(\n",
    "        alt.X(\"number of dilutions\", scale=alt.Scale(nice=False, padding=4)),\n",
    "        alt.Y(\"strain\", title=None),\n",
    "        alt.Column(\n",
    "            \"serum_replicate\",\n",
    "            title=None,\n",
    "            header=alt.Header(labelFontSize=12, labelFontStyle=\"bold\", labelPadding=0),\n",
    "        ),\n",
    "        alt.Fill(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['min_dilutions_per_barcode_serum_replicate']=}\",\n",
    "            legend=alt.Legend(titleLimit=500, orient=\"bottom\"),\n",
    "        ),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "        size=alt.condition(barcode_selection, alt.value(55), alt.value(35)),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if n_dilutions[c].dtype == float else c\n",
    "            for c in n_dilutions.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_circle(stroke=\"black\", strokeOpacity=1, fillOpacity=0.45)\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=120,\n",
    "        title=alt.TitleParams(\n",
    "            \"number of dilutions for each barcode for each serum-replicate\", dy=-2\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "display(n_dilutions_chart)\n",
    "\n",
    "# drop barcode / serum-replicates failing QC\n",
    "min_dilutions_per_barcode_serum_replicate_drops = list(\n",
    "    n_dilutions.query(\"fails_qc\")[[\"barcode\", \"serum_replicate\"]].itertuples(\n",
    "        index=False, name=None\n",
    "    )\n",
    ")\n",
    "print(\n",
    "    f\"\\nDropping {len(min_dilutions_per_barcode_serum_replicate_drops)} barcode/serum-replicates for failing \"\n",
    "    f\"{qc_thresholds['min_dilutions_per_barcode_serum_replicate']=}: \"\n",
    "    + str(min_dilutions_per_barcode_serum_replicate_drops)\n",
    ")\n",
    "qc_drops[\"barcode_serum_replicates\"].update(\n",
    "    {\n",
    "        w: \"min_dilutions_per_barcode_serum_replicate\"\n",
    "        for w in min_dilutions_per_barcode_serum_replicate_drops\n",
    "    }\n",
    ")\n",
    "frac_infectivity = frac_infectivity[\n",
    "    ~frac_infectivity.assign(\n",
    "        barcode_serum_replicate=lambda x: x.apply(\n",
    "            lambda r: (r[\"barcode\"], r[\"serum_replicate\"]), axis=1\n",
    "        )\n",
    "    )[\"barcode_serum_replicate\"].isin(qc_drops[\"barcode_serum_replicates\"])\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8822240e-5846-48da-ad31-53adaf647910",
   "metadata": {},
   "source": [
    "## Fit neutralization curves without applying QC to curves\n",
    "First fit curves to all serum replicates, then we will apply QC on the curve fits.\n",
    "Note that the fitting is done to the fraction infectivities **with** the ceiling:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51174a6b-e023-4dc3-812a-25e54b8b73f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fits_noqc = neutcurve.CurveFits(\n",
    "    frac_infectivity.rename(\n",
    "        columns={\n",
    "            \"frac_infectivity_ceiling\": \"fraction infectivity\",\n",
    "            \"concentration\": \"serum concentration\",\n",
    "        }\n",
    "    ),\n",
    "    conc_col=\"serum concentration\",\n",
    "    fracinf_col=\"fraction infectivity\",\n",
    "    virus_col=\"strain\",\n",
    "    serum_col=\"serum_replicate\",\n",
    "    replicate_col=\"barcode\",\n",
    "    fixtop=curvefit_params[\"fixtop\"],\n",
    "    fixbottom=curvefit_params[\"fixbottom\"],\n",
    "    fixslope=curvefit_params[\"fixslope\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6844135c-7272-4fad-942e-c2509283598f",
   "metadata": {},
   "source": [
    "Determine which fits fail the curve fitting QC, and plot them.\n",
    "Note the plot indicates as failing QC any barcode / serum-replicate that fails, even if we are also specified to ignore the QC for that one (so it will not be removed later):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef02fbc0-b123-4e88-9d54-97caafb21854",
   "metadata": {},
   "outputs": [],
   "source": [
    "goodness_of_fit = curvefit_qc[\"goodness_of_fit\"]\n",
    "\n",
    "fit_params_noqc = (\n",
    "    frac_infectivity.groupby([\"serum_replicate\", \"barcode\"], as_index=False)\n",
    "    .aggregate(max_frac_infectivity=pd.NamedAgg(\"frac_infectivity_ceiling\", \"max\"))\n",
    "    .merge(\n",
    "        fits_noqc.fitParams(average_only=False, no_average=True)[\n",
    "            [\"serum\", \"virus\", \"replicate\", \"r2\", \"rmsd\"]\n",
    "        ].rename(columns={\"serum\": \"serum_replicate\", \"replicate\": \"barcode\"}),\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    "    .assign(\n",
    "        fails_max_frac_infectivity_at_least=lambda x: (\n",
    "            x[\"max_frac_infectivity\"] < curvefit_qc[\"max_frac_infectivity_at_least\"]\n",
    "        ),\n",
    "        fails_goodness_of_fit=lambda x: (\n",
    "            (x[\"r2\"] < goodness_of_fit[\"min_R2\"])\n",
    "            & (x[\"rmsd\"] > goodness_of_fit[\"max_RMSD\"])\n",
    "        ),\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"fails_max_frac_infectivity_at_least\"] | x[\"fails_goodness_of_fit\"]\n",
    "        ),\n",
    "        ignore_qc=lambda x: x.apply(\n",
    "            lambda r: (\n",
    "                (\n",
    "                    r[\"serum_replicate\"]\n",
    "                    in curvefit_qc[\"serum_replicates_ignore_curvefit_qc\"]\n",
    "                )\n",
    "                or (\n",
    "                    (r[\"barcode\"], r[\"serum_replicate\"])\n",
    "                    in curvefit_qc[\"barcode_serum_replicates_ignore_curvefit_qc\"]\n",
    "                )\n",
    "            ),\n",
    "            axis=1,\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "print(f\"Plotting barcode / serum-replicates that fail {curvefit_qc=}\\n\")\n",
    "\n",
    "for prop, col in [\n",
    "    (\"max frac infectivity\", \"max_frac_infectivity\"),\n",
    "    (\"curve fit R2\", \"r2\"),\n",
    "    (\"curve fit RMSD\", \"rmsd\"),\n",
    "]:\n",
    "    fit_params_noqc_chart = (\n",
    "        alt.Chart(fit_params_noqc)\n",
    "        .add_params(barcode_selection)\n",
    "        .encode(\n",
    "            alt.X(col, title=prop, scale=alt.Scale(nice=False, padding=4)),\n",
    "            alt.Y(\"virus\", title=None),\n",
    "            alt.Fill(\"fails_qc\"),\n",
    "            alt.Column(\n",
    "                \"serum_replicate\",\n",
    "                title=None,\n",
    "                header=alt.Header(\n",
    "                    labelFontSize=12, labelFontStyle=\"bold\", labelPadding=0\n",
    "                ),\n",
    "            ),\n",
    "            strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "            size=alt.condition(barcode_selection, alt.value(55), alt.value(35)),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3g\") if fit_params_noqc[c].dtype == float else c\n",
    "                for c in fit_params_noqc.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_circle(stroke=\"black\", strokeOpacity=1, fillOpacity=0.55)\n",
    "        .properties(\n",
    "            height=alt.Step(10),\n",
    "            width=120,\n",
    "            title=alt.TitleParams(f\"{prop} for each barcode serum-replicate\", dy=-2),\n",
    "        )\n",
    "    )\n",
    "    display(fit_params_noqc_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5abc0ddb-ac7e-433e-8a2e-c1dd9af423a3",
   "metadata": {},
   "source": [
    "Now get all barcode / serum-replicate pairs that fail any of the QC.\n",
    "Plot curves for just these virus / serum-replicates (we plot all barcodes for a virus even if just one fails QC), and then exclude any that are not specified to ignore the QC:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7fa24aa-8596-48af-b868-20e2e6af84bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_serum_replicates_fail_qc = fit_params_noqc.query(\"fails_qc\").reset_index(\n",
    "    drop=True\n",
    ")\n",
    "print(f\"Here are barcode / serum-replicates that fail {curvefit_qc=}\")\n",
    "display(barcode_serum_replicates_fail_qc)\n",
    "\n",
    "if len(barcode_serum_replicates_fail_qc):\n",
    "    print(\"\\nCurves for viruses and serum-replicates with at least one failed barcode:\")\n",
    "    fig, _ = fits_noqc.plotReplicates(\n",
    "        sera=sorted(barcode_serum_replicates_fail_qc[\"serum_replicate\"].unique()),\n",
    "        viruses=sorted(barcode_serum_replicates_fail_qc[\"virus\"].unique()),\n",
    "        attempt_shared_legend=False,\n",
    "        legendfontsize=8,\n",
    "        titlesize=10,\n",
    "        ticksize=10,\n",
    "        ncol=6,\n",
    "        draw_in_bounds=True,\n",
    "    )\n",
    "    display(fig)\n",
    "    plt.close(fig)\n",
    "\n",
    "# drop barcode / serum-replicates failing QC\n",
    "for qc_filter in [\"max_frac_infectivity_at_least\", \"goodness_of_fit\"]:\n",
    "    fits_qc_drops = list(\n",
    "        fit_params_noqc.query(f\"fails_{qc_filter} and (not ignore_qc)\")[\n",
    "            [\"barcode\", \"serum_replicate\"]\n",
    "        ].itertuples(index=False, name=None)\n",
    "    )\n",
    "    print(\n",
    "        f\"\\nDropping {len(fits_qc_drops)} barcode/serum-replicates for failing \"\n",
    "        f\"{qc_filter}={curvefit_qc[qc_filter]}: \" + str(fits_qc_drops)\n",
    "    )\n",
    "    qc_drops[\"barcode_serum_replicates\"].update({w: qc_filter for w in fits_qc_drops})\n",
    "    frac_infectivity = frac_infectivity[\n",
    "        ~frac_infectivity.assign(\n",
    "            barcode_serum_replicate=lambda x: x.apply(\n",
    "                lambda r: (r[\"barcode\"], r[\"serum_replicate\"]), axis=1\n",
    "            )\n",
    "        )[\"barcode_serum_replicate\"].isin(qc_drops[\"barcode_serum_replicates\"])\n",
    "    ]\n",
    "    fit_params_noqc = fit_params_noqc[\n",
    "        ~fit_params_noqc.assign(\n",
    "            barcode_serum_replicate=lambda x: x.apply(\n",
    "                lambda r: (r[\"barcode\"], r[\"serum_replicate\"]), axis=1\n",
    "            )\n",
    "        )[\"barcode_serum_replicate\"].isin(qc_drops[\"barcode_serum_replicates\"])\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c45a632c-799e-4bf5-b6e9-da84d9ad6a4a",
   "metadata": {},
   "source": [
    "## Fit neutralization curves after applying QC\n",
    "No we re-fit curves after applying all the QC:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e8ed42-76be-474e-8e84-9a1765120b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fits_qc = neutcurve.CurveFits(\n",
    "    frac_infectivity.rename(\n",
    "        columns={\n",
    "            \"frac_infectivity_ceiling\": \"fraction infectivity\",\n",
    "            \"concentration\": \"serum concentration\",\n",
    "        }\n",
    "    ),\n",
    "    conc_col=\"serum concentration\",\n",
    "    fracinf_col=\"fraction infectivity\",\n",
    "    virus_col=\"strain\",\n",
    "    serum_col=\"serum\",\n",
    "    replicate_col=\"plate_barcode\",\n",
    "    fixtop=curvefit_params[\"fixtop\"],\n",
    "    fixbottom=curvefit_params[\"fixbottom\"],\n",
    "    fixslope=curvefit_params[\"fixslope\"],\n",
    ")\n",
    "\n",
    "fit_params_qc = fits_qc.fitParams(average_only=False, no_average=True)\n",
    "assert len(fit_params_qc) <= len(\n",
    "    fits_noqc.fitParams(average_only=False, no_average=True)\n",
    ")\n",
    "\n",
    "print(f\"Assigning fits for this plate to {group}\")\n",
    "fit_params_qc.insert(0, \"group\", group)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efaf94bb-7c12-4b56-bc2c-aced53738400",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-29T21:30:37.367518Z",
     "iopub.status.busy": "2023-12-29T21:30:37.366746Z",
     "iopub.status.idle": "2023-12-29T21:30:37.391447Z",
     "shell.execute_reply": "2023-12-29T21:30:37.390610Z",
     "shell.execute_reply.started": "2023-12-29T21:30:37.367449Z"
    }
   },
   "source": [
    "Plot all the curves that passed QC:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259d7ff1-ecfb-46ed-acac-4b38f6587bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fits_qc.sera:\n",
    "    _ = fits_qc.plotReplicates(\n",
    "        attempt_shared_legend=False,\n",
    "        legendfontsize=8,\n",
    "        titlesize=10,\n",
    "        ticksize=10,\n",
    "        ncol=6,\n",
    "        draw_in_bounds=True,\n",
    "    )\n",
    "else:\n",
    "    print(\"No sera passed QC.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f0832bd-4632-4954-ae9d-ac65625ccc83",
   "metadata": {},
   "source": [
    "## Save results to files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1773efda-c871-4b10-a6b8-f192cac6991e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Writing fraction infectivities to {frac_infectivity_csv}\")\n",
    "(\n",
    "    frac_infectivity[\n",
    "        [\n",
    "            \"serum\",\n",
    "            \"strain\",\n",
    "            \"plate_barcode\",\n",
    "            \"dilution_factor\",\n",
    "            \"frac_infectivity_raw\",\n",
    "            \"frac_infectivity_ceiling\",\n",
    "        ]\n",
    "    ]\n",
    "    .sort_values([\"serum\", \"plate_barcode\", \"dilution_factor\"])\n",
    "    .to_csv(frac_infectivity_csv, index=False, float_format=\"%.4g\")\n",
    ")\n",
    "\n",
    "print(f\"\\nWriting fit parameters to {fits_csv}\")\n",
    "(\n",
    "    fit_params_qc.drop(columns=[\"nreplicates\", \"ic50_str\"]).to_csv(\n",
    "        fits_csv, index=False, float_format=\"%.4g\"\n",
    "    )\n",
    ")\n",
    "\n",
    "print(f\"\\nPickling neutcurve.CurveFits object for these data to {fits_pickle}\")\n",
    "with open(fits_pickle, \"wb\") as f:\n",
    "    pickle.dump(fits_qc, f)\n",
    "\n",
    "print(f\"\\nWriting QC drops to {qc_drops_yaml}\")\n",
    "\n",
    "\n",
    "def tup_to_str(x):\n",
    "    return \" \".join(x) if isinstance(x, tuple) else x\n",
    "\n",
    "\n",
    "qc_drops_for_yaml = {\n",
    "    key: {tup_to_str(key2): val2 for key2, val2 in val.items()}\n",
    "    for key, val in qc_drops.items()\n",
    "}\n",
    "with open(qc_drops_yaml, \"w\") as f:\n",
    "    yaml.YAML(typ=\"rt\").dump(qc_drops_for_yaml, f)\n",
    "print(\"\\nHere are the QC drops:\\n***************************\")\n",
    "yaml.YAML(typ=\"rt\").dump(qc_drops_for_yaml, sys.stdout)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
