{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4eb38ea4-98fa-4da1-9071-2c65d38291ff",
   "metadata": {},
   "source": [
    "# Process plate counts to get fraction infectivities and fit curves\n",
    "This notebook is designed to be run using `snakemake`, and analyzes a plate of sequencing-based neutralization assays.\n",
    "\n",
    "The plots generated by this notebook are interactive, so you can mouseover points for details, use the mouse-scroll to zoom and pan, and use interactive dropdowns at the bottom of the plots.\n",
    "\n",
    "## Setup\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "900fe170-1b4b-4646-b9ee-01d75a76f664",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.081523Z",
     "iopub.status.busy": "2023-12-28T14:34:21.080856Z",
     "iopub.status.idle": "2023-12-28T14:34:21.090052Z",
     "shell.execute_reply": "2023-12-28T14:34:21.088674Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.081459Z"
    }
   },
   "outputs": [],
   "source": [
    "import collections\n",
    "import sys\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import neutcurve\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import ruamel.yaml as yaml\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc47332-2e29-4605-bf8b-463edb63cac6",
   "metadata": {},
   "source": [
    "Get the variables passed by `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e05d899b-01b3-4e9a-8b80-6c3414434a31",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.138788Z",
     "iopub.status.busy": "2023-12-28T14:34:21.138331Z",
     "iopub.status.idle": "2023-12-28T14:34:21.165675Z",
     "shell.execute_reply": "2023-12-28T14:34:21.164644Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.138753Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/fh/fast/bloom_j/software/miniconda3/envs/seqneut-pipeline/lib/python3.11/site-packages', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline/test_example/..', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline/test_example', '/fh/fast/bloom_j/software/miniconda3/envs/seqneut-pipeline/bin', '/fh/fast/bloom_j/software/miniconda3/envs/seqneut-pipeline/lib/python3.11', '/fh/fast/bloom_j/software/miniconda3/envs/seqneut-pipeline/lib/python3.11/lib-dynload', '/home/jbloom/.local/lib/python3.11/site-packages', '/fh/fast/bloom_j/software/miniconda3/envs/seqneut-pipeline/lib/python3.11/site-packages', '/home/jbloom/.cache/snakemake/snakemake/source-cache/runtime-cache/tmp8ojbjvb9/file/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline/notebooks', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95a5\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c*results/barcode_counts/plate11_none-10.csv\\x94\\x8c-results/barcode_counts/plate11_M099d30_20.csv\\x94\\x8c-results/barcode_counts/plate11_M099d30_60.csv\\x94\\x8c.results/barcode_counts/plate11_M099d30_180.csv\\x94\\x8c.results/barcode_counts/plate11_M099d30_540.csv\\x94\\x8c/results/barcode_counts/plate11_M099d30_1620.csv\\x94\\x8c/results/barcode_counts/plate11_M099d30_4860.csv\\x94\\x8c0results/barcode_counts/plate11_M099d30_14580.csv\\x94\\x8c0results/barcode_counts/plate11_M099d30_43740.csv\\x94\\x8c1results/barcode_counts/plate11_M099d30_131220.csv\\x94\\x8c1results/barcode_counts/plate11_M099d30_393660.csv\\x94\\x8c)results/barcode_counts/plate11_none-2.csv\\x94\\x8c*results/barcode_counts/plate11_none-11.csv\\x94\\x8c,results/barcode_counts/plate11_M099d0_20.csv\\x94\\x8c,results/barcode_counts/plate11_M099d0_60.csv\\x94\\x8c-results/barcode_counts/plate11_M099d0_180.csv\\x94\\x8c-results/barcode_counts/plate11_M099d0_540.csv\\x94\\x8c.results/barcode_counts/plate11_M099d0_1620.csv\\x94\\x8c.results/barcode_counts/plate11_M099d0_4860.csv\\x94\\x8c/results/barcode_counts/plate11_M099d0_14580.csv\\x94\\x8c/results/barcode_counts/plate11_M099d0_43740.csv\\x94\\x8c0results/barcode_counts/plate11_M099d0_131220.csv\\x94\\x8c0results/barcode_counts/plate11_M099d0_393660.csv\\x94\\x8c)results/barcode_counts/plate11_none-3.csv\\x94\\x8c*results/barcode_counts/plate11_none-12.csv\\x94\\x8c-results/barcode_counts/plate11_Y044d30_20.csv\\x94\\x8c-results/barcode_counts/plate11_Y044d30_60.csv\\x94\\x8c.results/barcode_counts/plate11_Y044d30_180.csv\\x94\\x8c.results/barcode_counts/plate11_Y044d30_540.csv\\x94\\x8c/results/barcode_counts/plate11_Y044d30_1620.csv\\x94\\x8c/results/barcode_counts/plate11_Y044d30_4860.csv\\x94\\x8c0results/barcode_counts/plate11_Y044d30_14580.csv\\x94\\x8c0results/barcode_counts/plate11_Y044d30_43740.csv\\x94\\x8c1results/barcode_counts/plate11_Y044d30_131220.csv\\x94\\x8c1results/barcode_counts/plate11_Y044d30_393660.csv\\x94\\x8c)results/barcode_counts/plate11_none-4.csv\\x94\\x8c)results/barcode_fates/plate11_none-10.csv\\x94\\x8c,results/barcode_fates/plate11_M099d30_20.csv\\x94\\x8c,results/barcode_fates/plate11_M099d30_60.csv\\x94\\x8c-results/barcode_fates/plate11_M099d30_180.csv\\x94\\x8c-results/barcode_fates/plate11_M099d30_540.csv\\x94\\x8c.results/barcode_fates/plate11_M099d30_1620.csv\\x94\\x8c.results/barcode_fates/plate11_M099d30_4860.csv\\x94\\x8c/results/barcode_fates/plate11_M099d30_14580.csv\\x94\\x8c/results/barcode_fates/plate11_M099d30_43740.csv\\x94\\x8c0results/barcode_fates/plate11_M099d30_131220.csv\\x94\\x8c0results/barcode_fates/plate11_M099d30_393660.csv\\x94\\x8c(results/barcode_fates/plate11_none-2.csv\\x94\\x8c)results/barcode_fates/plate11_none-11.csv\\x94\\x8c+results/barcode_fates/plate11_M099d0_20.csv\\x94\\x8c+results/barcode_fates/plate11_M099d0_60.csv\\x94\\x8c,results/barcode_fates/plate11_M099d0_180.csv\\x94\\x8c,results/barcode_fates/plate11_M099d0_540.csv\\x94\\x8c-results/barcode_fates/plate11_M099d0_1620.csv\\x94\\x8c-results/barcode_fates/plate11_M099d0_4860.csv\\x94\\x8c.results/barcode_fates/plate11_M099d0_14580.csv\\x94\\x8c.results/barcode_fates/plate11_M099d0_43740.csv\\x94\\x8c/results/barcode_fates/plate11_M099d0_131220.csv\\x94\\x8c/results/barcode_fates/plate11_M099d0_393660.csv\\x94\\x8c(results/barcode_fates/plate11_none-3.csv\\x94\\x8c)results/barcode_fates/plate11_none-12.csv\\x94\\x8c,results/barcode_fates/plate11_Y044d30_20.csv\\x94\\x8c,results/barcode_fates/plate11_Y044d30_60.csv\\x94\\x8c-results/barcode_fates/plate11_Y044d30_180.csv\\x94\\x8c-results/barcode_fates/plate11_Y044d30_540.csv\\x94\\x8c.results/barcode_fates/plate11_Y044d30_1620.csv\\x94\\x8c.results/barcode_fates/plate11_Y044d30_4860.csv\\x94\\x8c/results/barcode_fates/plate11_Y044d30_14580.csv\\x94\\x8c/results/barcode_fates/plate11_Y044d30_43740.csv\\x94\\x8c0results/barcode_fates/plate11_Y044d30_131220.csv\\x94\\x8c0results/barcode_fates/plate11_Y044d30_393660.csv\\x94\\x8c(results/barcode_fates/plate11_none-4.csv\\x94\\x8c-data/viral_libraries/pdmH1N1_lib2023_loes.csv\\x94\\x8c3data/neut_standard_sets/loes2023_neut_standards.csv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\ncount_csvs\\x94K\\x00K$\\x86\\x94\\x8c\\tfate_csvs\\x94K$KH\\x86\\x94\\x8c\\x11viral_library_csv\\x94KHN\\x86\\x94\\x8c\\x15neut_standard_set_csv\\x94KIN\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eha\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(hg)}\\x94\\x8c\\x05_name\\x94hasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bhWh\\x06\\x8c\\tNamedlist\\x94\\x93\\x94)\\x81\\x94(h\\nh\\x0bh\\x0ch\\rh\\x0eh\\x0fh\\x10h\\x11h\\x12h\\x13h\\x14h\\x15h\\x16h\\x17h\\x18h\\x19h\\x1ah\\x1bh\\x1ch\\x1dh\\x1eh\\x1fh h!h\"h#h$h%h&h\\'h(h)h*h+h,h-e}\\x94(hU}\\x94h_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bubhYhr)\\x81\\x94(h.h/h0h1h2h3h4h5h6h7h8h9h:h;h<h=h>h?h@hAhBhChDhEhFhGhHhIhJhKhLhMhNhOhPhQe}\\x94(hU}\\x94h_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bubh[hRh]hSub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c5results/plates/plate11/process_counts_qc_failures.yml\\x94\\x8c+results/plates/plate11/frac_infectivity.csv\\x94\\x8c$results/plates/plate11/curvefits.csv\\x94\\x8c\\'results/plates/plate11/curvefits.pickle\\x94e}\\x94(hU}\\x94(\\x8c\\x0bqc_failures\\x94K\\x00N\\x86\\x94\\x8c\\x14frac_infectivity_csv\\x94K\\x01N\\x86\\x94\\x8c\\x08fits_csv\\x94K\\x02N\\x86\\x94\\x8c\\x0bfits_pickle\\x94K\\x03N\\x86\\x94uh_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bh\\x95h\\x8fh\\x97h\\x90h\\x99h\\x91h\\x9bh\\x92ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94(]\\x94(\\x8c\\x0fplate11_none-10\\x94\\x8c\\x12plate11_M099d30_20\\x94\\x8c\\x12plate11_M099d30_60\\x94\\x8c\\x13plate11_M099d30_180\\x94\\x8c\\x13plate11_M099d30_540\\x94\\x8c\\x14plate11_M099d30_1620\\x94\\x8c\\x14plate11_M099d30_4860\\x94\\x8c\\x15plate11_M099d30_14580\\x94\\x8c\\x15plate11_M099d30_43740\\x94\\x8c\\x16plate11_M099d30_131220\\x94\\x8c\\x16plate11_M099d30_393660\\x94\\x8c\\x0eplate11_none-2\\x94\\x8c\\x0fplate11_none-11\\x94\\x8c\\x11plate11_M099d0_20\\x94\\x8c\\x11plate11_M099d0_60\\x94\\x8c\\x12plate11_M099d0_180\\x94\\x8c\\x12plate11_M099d0_540\\x94\\x8c\\x13plate11_M099d0_1620\\x94\\x8c\\x13plate11_M099d0_4860\\x94\\x8c\\x14plate11_M099d0_14580\\x94\\x8c\\x14plate11_M099d0_43740\\x94\\x8c\\x15plate11_M099d0_131220\\x94\\x8c\\x15plate11_M099d0_393660\\x94\\x8c\\x0eplate11_none-3\\x94\\x8c\\x0fplate11_none-12\\x94\\x8c\\x12plate11_Y044d30_20\\x94\\x8c\\x12plate11_Y044d30_60\\x94\\x8c\\x13plate11_Y044d30_180\\x94\\x8c\\x13plate11_Y044d30_540\\x94\\x8c\\x14plate11_Y044d30_1620\\x94\\x8c\\x14plate11_Y044d30_4860\\x94\\x8c\\x15plate11_Y044d30_14580\\x94\\x8c\\x15plate11_Y044d30_43740\\x94\\x8c\\x16plate11_Y044d30_131220\\x94\\x8c\\x16plate11_Y044d30_393660\\x94\\x8c\\x0eplate11_none-4\\x94e}\\x94(\\x8c\\x04date\\x94\\x8c\\n2023-09-26\\x94\\x8c\\rviral_library\\x94\\x8c\\x14pdmH1N1_lib2023_loes\\x94\\x8c\\x11neut_standard_set\\x94\\x8c\\x08loes2023\\x94\\x8c\\x0bsamples_csv\\x94\\x8c\\x1fdata/plates/plate11_samples.csv\\x94\\x8c\\rqc_thresholds\\x94}\\x94(\\x8c\\x1bavg_barcode_counts_per_well\\x94K\\xfa\\x8c\\x1fmin_neut_standard_frac_per_well\\x94G?tz\\xe1G\\xae\\x14{\\x8c\\x1cno_serum_barcode_consistency\\x94}\\x94(\\x8c\\x0fmax_fold_change\\x94K\\x03\\x8c\\tmax_wells\\x94K\\x02u\\x8c!neut_standard_barcode_consistency\\x94}\\x94(\\x8c\\x0fmax_fold_change\\x94K\\x03\\x8c\\x0cmax_barcodes\\x94K\\x02u\\x8c\\x18barcode_frac_consistency\\x94K\\x04\\x8c\\x16min_viral_barcode_frac\\x94G?@bM\\xd2\\xf1\\xa9\\xfc\\x8c\\x1emin_neut_standard_barcode_frac\\x94G?tz\\xe1G\\xae\\x14{\\x8c\\x17min_neut_standard_count\\x94M^\\x01\\x8c min_no_serum_viral_barcode_count\\x94K\\x05\\x8c!min_dilutions_per_serum_replicate\\x94K\\x04\\x8c\\x14max_frac_infectivity\\x94K\\x08u\\x8c\\x10barcodes_to_drop\\x94]\\x94\\x8c\\x10GGTCCATCTCAGATCG\\x94a\\x8c\\rwells_to_drop\\x94]\\x94(\\x8c\\x02D6\\x94\\x8c\\x03C12\\x94e\\x8c\\x0fcurvefit_params\\x94}\\x94(\\x8c\\x18frac_infectivity_ceiling\\x94K\\x01\\x8c\\x06fixtop\\x94\\x89\\x8c\\tfixbottom\\x94K\\x00u\\x8c\\x07samples\\x94}\\x94(\\x8c\\x04well\\x94}\\x94(K\\x00\\x8c\\x02B1\\x94K\\x01\\x8c\\x02B2\\x94K\\x02\\x8c\\x02B3\\x94K\\x03\\x8c\\x02B4\\x94K\\x04\\x8c\\x02B5\\x94K\\x05\\x8c\\x02B6\\x94K\\x06\\x8c\\x02B7\\x94K\\x07\\x8c\\x02B8\\x94K\\x08\\x8c\\x02B9\\x94K\\t\\x8c\\x03B10\\x94K\\n\\x8c\\x03B11\\x94K\\x0b\\x8c\\x03B12\\x94K\\x0c\\x8c\\x02C1\\x94K\\r\\x8c\\x02C2\\x94K\\x0e\\x8c\\x02C3\\x94K\\x0f\\x8c\\x02C4\\x94K\\x10\\x8c\\x02C5\\x94K\\x11\\x8c\\x02C6\\x94K\\x12\\x8c\\x02C7\\x94K\\x13\\x8c\\x02C8\\x94K\\x14\\x8c\\x02C9\\x94K\\x15\\x8c\\x03C10\\x94K\\x16\\x8c\\x03C11\\x94K\\x17\\x8c\\x03C12\\x94K\\x18\\x8c\\x02D1\\x94K\\x19\\x8c\\x02D2\\x94K\\x1a\\x8c\\x02D3\\x94K\\x1b\\x8c\\x02D4\\x94K\\x1c\\x8c\\x02D5\\x94K\\x1d\\x8c\\x02D6\\x94K\\x1e\\x8c\\x02D7\\x94K\\x1f\\x8c\\x02D8\\x94K \\x8c\\x02D9\\x94K!\\x8c\\x03D10\\x94K\"\\x8c\\x03D11\\x94K#\\x8c\\x03D12\\x94u\\x8c\\x05serum\\x94}\\x94(K\\x00\\x8c\\x04none\\x94K\\x01\\x8c\\x07M099d30\\x94K\\x02j\"\\x01\\x00\\x00K\\x03j\"\\x01\\x00\\x00K\\x04j\"\\x01\\x00\\x00K\\x05j\"\\x01\\x00\\x00K\\x06j\"\\x01\\x00\\x00K\\x07j\"\\x01\\x00\\x00K\\x08j\"\\x01\\x00\\x00K\\tj\"\\x01\\x00\\x00K\\nj\"\\x01\\x00\\x00K\\x0bj!\\x01\\x00\\x00K\\x0cj!\\x01\\x00\\x00K\\r\\x8c\\x06M099d0\\x94K\\x0ej#\\x01\\x00\\x00K\\x0fj#\\x01\\x00\\x00K\\x10j#\\x01\\x00\\x00K\\x11j#\\x01\\x00\\x00K\\x12j#\\x01\\x00\\x00K\\x13j#\\x01\\x00\\x00K\\x14j#\\x01\\x00\\x00K\\x15j#\\x01\\x00\\x00K\\x16j#\\x01\\x00\\x00K\\x17j!\\x01\\x00\\x00K\\x18j!\\x01\\x00\\x00K\\x19\\x8c\\x07Y044d30\\x94K\\x1aj$\\x01\\x00\\x00K\\x1bj$\\x01\\x00\\x00K\\x1cj$\\x01\\x00\\x00K\\x1dj$\\x01\\x00\\x00K\\x1ej$\\x01\\x00\\x00K\\x1fj$\\x01\\x00\\x00K j$\\x01\\x00\\x00K!j$\\x01\\x00\\x00K\"j$\\x01\\x00\\x00K#j!\\x01\\x00\\x00u\\x8c\\x0fdilution_factor\\x94}\\x94(K\\x00NK\\x01K\\x14K\\x02K<K\\x03K\\xb4K\\x04M\\x1c\\x02K\\x05MT\\x06K\\x06M\\xfc\\x12K\\x07M\\xf48K\\x08M\\xdc\\xaaK\\tJ\\x94\\x00\\x02\\x00K\\nJ\\xbc\\x01\\x06\\x00K\\x0bNK\\x0cNK\\rK\\x14K\\x0eK<K\\x0fK\\xb4K\\x10M\\x1c\\x02K\\x11MT\\x06K\\x12M\\xfc\\x12K\\x13M\\xf48K\\x14M\\xdc\\xaaK\\x15J\\x94\\x00\\x02\\x00K\\x16J\\xbc\\x01\\x06\\x00K\\x17NK\\x18NK\\x19K\\x14K\\x1aK<K\\x1bK\\xb4K\\x1cM\\x1c\\x02K\\x1dMT\\x06K\\x1eM\\xfc\\x12K\\x1fM\\xf48K M\\xdc\\xaaK!J\\x94\\x00\\x02\\x00K\"J\\xbc\\x01\\x06\\x00K#Nu\\x8c\\treplicate\\x94}\\x94(K\\x00K\\nK\\x01K\\x02K\\x02K\\x02K\\x03K\\x02K\\x04K\\x02K\\x05K\\x02K\\x06K\\x02K\\x07K\\x02K\\x08K\\x02K\\tK\\x02K\\nK\\x02K\\x0bK\\x02K\\x0cK\\x0bK\\rK\\x02K\\x0eK\\x02K\\x0fK\\x02K\\x10K\\x02K\\x11K\\x02K\\x12K\\x02K\\x13K\\x02K\\x14K\\x02K\\x15K\\x02K\\x16K\\x02K\\x17K\\x03K\\x18K\\x0cK\\x19K\\x02K\\x1aK\\x02K\\x1bK\\x02K\\x1cK\\x02K\\x1dK\\x02K\\x1eK\\x02K\\x1fK\\x02K K\\x02K!K\\x02K\"K\\x02K#K\\x04u\\x8c\\x05fastq\\x94}\\x94(K\\x00\\x8c*fastqs/Plate11_Noserum2_S2_R1_001.fastq.gz\\x94K\\x01\\x8c)fastqs/M099_d30_conc1_S10_R1_001.fastq.gz\\x94K\\x02\\x8c)fastqs/M099_d30_conc2_S18_R1_001.fastq.gz\\x94K\\x03\\x8c)fastqs/M099_d30_conc3_S26_R1_001.fastq.gz\\x94K\\x04\\x8c)fastqs/M099_d30_conc4_S34_R1_001.fastq.gz\\x94K\\x05\\x8c)fastqs/M099_d30_conc5_S42_R1_001.fastq.gz\\x94K\\x06\\x8c)fastqs/M099_d30_conc6_S50_R1_001.fastq.gz\\x94K\\x07\\x8c)fastqs/M099_d30_conc7_S58_R1_001.fastq.gz\\x94K\\x08\\x8c)fastqs/M099_d30_conc8_S66_R1_001.fastq.gz\\x94K\\t\\x8c)fastqs/M099_d30_conc9_S74_R1_001.fastq.gz\\x94K\\n\\x8c*fastqs/M099_d30_conc10_S82_R1_001.fastq.gz\\x94K\\x0b\\x8c,fastqs/Plate11_Noserum10_S90_R1_001.fastq.gz\\x94K\\x0c\\x8c*fastqs/Plate11_Noserum3_S3_R1_001.fastq.gz\\x94K\\r\\x8c(fastqs/M099_d0_conc1_S11_R1_001.fastq.gz\\x94K\\x0e\\x8c(fastqs/M099_d0_conc2_S19_R1_001.fastq.gz\\x94K\\x0f\\x8c(fastqs/M099_d0_conc3_S27_R1_001.fastq.gz\\x94K\\x10\\x8c(fastqs/M099_d0_conc4_S35_R1_001.fastq.gz\\x94K\\x11\\x8c(fastqs/M099_d0_conc5_S43_R1_001.fastq.gz\\x94K\\x12\\x8c(fastqs/M099_d0_conc6_S51_R1_001.fastq.gz\\x94K\\x13\\x8c(fastqs/M099_d0_conc7_S59_R1_001.fastq.gz\\x94K\\x14\\x8c(fastqs/M099_d0_conc8_S67_R1_001.fastq.gz\\x94K\\x15\\x8c(fastqs/M099_d0_conc9_S75_R1_001.fastq.gz\\x94K\\x16\\x8c)fastqs/M099_d0_conc10_S83_R1_001.fastq.gz\\x94K\\x17\\x8c,fastqs/Plate11_Noserum11_S91_R1_001.fastq.gz\\x94K\\x18\\x8c*fastqs/Plate11_Noserum4_S4_R1_001.fastq.gz\\x94K\\x19\\x8c)fastqs/Y044_d30_conc1_S12_R1_001.fastq.gz\\x94K\\x1a\\x8c)fastqs/Y044_d30_conc2_S20_R1_001.fastq.gz\\x94K\\x1b\\x8c)fastqs/Y044_d30_conc3_S28_R1_001.fastq.gz\\x94K\\x1c\\x8c)fastqs/Y044_d30_conc4_S36_R1_001.fastq.gz\\x94K\\x1d\\x8c)fastqs/Y044_d30_conc5_S44_R1_001.fastq.gz\\x94K\\x1e\\x8c)fastqs/Y044_d30_conc6_S52_R1_001.fastq.gz\\x94K\\x1f\\x8c)fastqs/Y044_d30_conc7_S60_R1_001.fastq.gz\\x94K \\x8c)fastqs/Y044_d30_conc8_S68_R1_001.fastq.gz\\x94K!\\x8c)fastqs/Y044_d30_conc9_S76_R1_001.fastq.gz\\x94K\"\\x8c*fastqs/Y044_d30_conc10_S84_R1_001.fastq.gz\\x94K#\\x8c,fastqs/Plate11_Noserum12_S92_R1_001.fastq.gz\\x94u\\x8c\\x0fserum_replicate\\x94}\\x94(K\\x00\\x8c\\x07none-10\\x94K\\x01j\"\\x01\\x00\\x00K\\x02j\"\\x01\\x00\\x00K\\x03j\"\\x01\\x00\\x00K\\x04j\"\\x01\\x00\\x00K\\x05j\"\\x01\\x00\\x00K\\x06j\"\\x01\\x00\\x00K\\x07j\"\\x01\\x00\\x00K\\x08j\"\\x01\\x00\\x00K\\tj\"\\x01\\x00\\x00K\\nj\"\\x01\\x00\\x00K\\x0b\\x8c\\x06none-2\\x94K\\x0c\\x8c\\x07none-11\\x94K\\rj#\\x01\\x00\\x00K\\x0ej#\\x01\\x00\\x00K\\x0fj#\\x01\\x00\\x00K\\x10j#\\x01\\x00\\x00K\\x11j#\\x01\\x00\\x00K\\x12j#\\x01\\x00\\x00K\\x13j#\\x01\\x00\\x00K\\x14j#\\x01\\x00\\x00K\\x15j#\\x01\\x00\\x00K\\x16j#\\x01\\x00\\x00K\\x17\\x8c\\x06none-3\\x94K\\x18\\x8c\\x07none-12\\x94K\\x19j$\\x01\\x00\\x00K\\x1aj$\\x01\\x00\\x00K\\x1bj$\\x01\\x00\\x00K\\x1cj$\\x01\\x00\\x00K\\x1dj$\\x01\\x00\\x00K\\x1ej$\\x01\\x00\\x00K\\x1fj$\\x01\\x00\\x00K j$\\x01\\x00\\x00K!j$\\x01\\x00\\x00K\"j$\\x01\\x00\\x00K#\\x8c\\x06none-4\\x94u\\x8c\\x0esample_noplate\\x94}\\x94(K\\x00jQ\\x01\\x00\\x00K\\x01\\x8c\\nM099d30_20\\x94K\\x02\\x8c\\nM099d30_60\\x94K\\x03\\x8c\\x0bM099d30_180\\x94K\\x04\\x8c\\x0bM099d30_540\\x94K\\x05\\x8c\\x0cM099d30_1620\\x94K\\x06\\x8c\\x0cM099d30_4860\\x94K\\x07\\x8c\\rM099d30_14580\\x94K\\x08\\x8c\\rM099d30_43740\\x94K\\t\\x8c\\x0eM099d30_131220\\x94K\\n\\x8c\\x0eM099d30_393660\\x94K\\x0bjR\\x01\\x00\\x00K\\x0cjS\\x01\\x00\\x00K\\r\\x8c\\tM099d0_20\\x94K\\x0e\\x8c\\tM099d0_60\\x94K\\x0f\\x8c\\nM099d0_180\\x94K\\x10\\x8c\\nM099d0_540\\x94K\\x11\\x8c\\x0bM099d0_1620\\x94K\\x12\\x8c\\x0bM099d0_4860\\x94K\\x13\\x8c\\x0cM099d0_14580\\x94K\\x14\\x8c\\x0cM099d0_43740\\x94K\\x15\\x8c\\rM099d0_131220\\x94K\\x16\\x8c\\rM099d0_393660\\x94K\\x17jT\\x01\\x00\\x00K\\x18jU\\x01\\x00\\x00K\\x19\\x8c\\nY044d30_20\\x94K\\x1a\\x8c\\nY044d30_60\\x94K\\x1b\\x8c\\x0bY044d30_180\\x94K\\x1c\\x8c\\x0bY044d30_540\\x94K\\x1d\\x8c\\x0cY044d30_1620\\x94K\\x1e\\x8c\\x0cY044d30_4860\\x94K\\x1f\\x8c\\rY044d30_14580\\x94K \\x8c\\rY044d30_43740\\x94K!\\x8c\\x0eY044d30_131220\\x94K\"\\x8c\\x0eY044d30_393660\\x94K#jV\\x01\\x00\\x00u\\x8c\\x06sample\\x94}\\x94(K\\x00h\\xabK\\x01h\\xacK\\x02h\\xadK\\x03h\\xaeK\\x04h\\xafK\\x05h\\xb0K\\x06h\\xb1K\\x07h\\xb2K\\x08h\\xb3K\\th\\xb4K\\nh\\xb5K\\x0bh\\xb6K\\x0ch\\xb7K\\rh\\xb8K\\x0eh\\xb9K\\x0fh\\xbaK\\x10h\\xbbK\\x11h\\xbcK\\x12h\\xbdK\\x13h\\xbeK\\x14h\\xbfK\\x15h\\xc0K\\x16h\\xc1K\\x17h\\xc2K\\x18h\\xc3K\\x19h\\xc4K\\x1ah\\xc5K\\x1bh\\xc6K\\x1ch\\xc7K\\x1dh\\xc8K\\x1eh\\xc9K\\x1fh\\xcaK h\\xcbK!h\\xccK\"h\\xcdK#h\\xceu\\x8c\\x05plate\\x94}\\x94(K\\x00\\x8c\\x07plate11\\x94K\\x01j{\\x01\\x00\\x00K\\x02j{\\x01\\x00\\x00K\\x03j{\\x01\\x00\\x00K\\x04j{\\x01\\x00\\x00K\\x05j{\\x01\\x00\\x00K\\x06j{\\x01\\x00\\x00K\\x07j{\\x01\\x00\\x00K\\x08j{\\x01\\x00\\x00K\\tj{\\x01\\x00\\x00K\\nj{\\x01\\x00\\x00K\\x0bj{\\x01\\x00\\x00K\\x0cj{\\x01\\x00\\x00K\\rj{\\x01\\x00\\x00K\\x0ej{\\x01\\x00\\x00K\\x0fj{\\x01\\x00\\x00K\\x10j{\\x01\\x00\\x00K\\x11j{\\x01\\x00\\x00K\\x12j{\\x01\\x00\\x00K\\x13j{\\x01\\x00\\x00K\\x14j{\\x01\\x00\\x00K\\x15j{\\x01\\x00\\x00K\\x16j{\\x01\\x00\\x00K\\x17j{\\x01\\x00\\x00K\\x18j{\\x01\\x00\\x00K\\x19j{\\x01\\x00\\x00K\\x1aj{\\x01\\x00\\x00K\\x1bj{\\x01\\x00\\x00K\\x1cj{\\x01\\x00\\x00K\\x1dj{\\x01\\x00\\x00K\\x1ej{\\x01\\x00\\x00K\\x1fj{\\x01\\x00\\x00K j{\\x01\\x00\\x00K!j{\\x01\\x00\\x00K\"j{\\x01\\x00\\x00K#j{\\x01\\x00\\x00u\\x8c\\x0fplate_replicate\\x94}\\x94(K\\x00\\x8c\\nplate11-10\\x94K\\x01j{\\x01\\x00\\x00K\\x02j{\\x01\\x00\\x00K\\x03j{\\x01\\x00\\x00K\\x04j{\\x01\\x00\\x00K\\x05j{\\x01\\x00\\x00K\\x06j{\\x01\\x00\\x00K\\x07j{\\x01\\x00\\x00K\\x08j{\\x01\\x00\\x00K\\tj{\\x01\\x00\\x00K\\nj{\\x01\\x00\\x00K\\x0b\\x8c\\tplate11-2\\x94K\\x0c\\x8c\\nplate11-11\\x94K\\rj{\\x01\\x00\\x00K\\x0ej{\\x01\\x00\\x00K\\x0fj{\\x01\\x00\\x00K\\x10j{\\x01\\x00\\x00K\\x11j{\\x01\\x00\\x00K\\x12j{\\x01\\x00\\x00K\\x13j{\\x01\\x00\\x00K\\x14j{\\x01\\x00\\x00K\\x15j{\\x01\\x00\\x00K\\x16j{\\x01\\x00\\x00K\\x17\\x8c\\tplate11-3\\x94K\\x18\\x8c\\nplate11-12\\x94K\\x19j{\\x01\\x00\\x00K\\x1aj{\\x01\\x00\\x00K\\x1bj{\\x01\\x00\\x00K\\x1cj{\\x01\\x00\\x00K\\x1dj{\\x01\\x00\\x00K\\x1ej{\\x01\\x00\\x00K\\x1fj{\\x01\\x00\\x00K j{\\x01\\x00\\x00K!j{\\x01\\x00\\x00K\"j{\\x01\\x00\\x00K#\\x8c\\tplate11-4\\x94uuue}\\x94(hU}\\x94(h\\xf7K\\x00N\\x86\\x94\\x8c\\x0cplate_params\\x94K\\x01N\\x86\\x94uh_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bh\\xf7h\\xaaj\\x87\\x01\\x00\\x00h\\xcfub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\x07plate11\\x94a}\\x94(hU}\\x94\\x8c\\x05plate\\x94K\\x00N\\x86\\x94sh_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bjy\\x01\\x00\\x00j\\x96\\x01\\x00\\x00ub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(hU}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bj\\xac\\x01\\x00\\x00K\\x01j\\xae\\x01\\x00\\x00K\\x01j\\xb0\\x01\\x00\\x00j\\xa9\\x01\\x00\\x00ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c,results/plates/plate11/process_plate11.ipynb\\x94a}\\x94(hU}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh_]\\x94(hahbehahehg\\x85\\x94R\\x94(hg)}\\x94hkhasNt\\x94bhbhehg\\x85\\x94R\\x94(hg)}\\x94hkhbsNt\\x94bj\\xc2\\x01\\x00\\x00j\\xbf\\x01\\x00\\x00ub\\x8c\\x06config\\x94}\\x94(\\x8c\\x10seqneut-pipeline\\x94\\x8c\\x03../\\x94\\x8c\\x04docs\\x94\\x8c\\x07../docs\\x94\\x8c\\x0bdescription\\x94X\\xba\\x01\\x00\\x00# Test example for [seqneut-pipeline](https://github.com/jbloomlab/seqneut-pipeline)\\nThis is a small toy-example created by subsetting a real experiment dataset.\\n\\nSee [https://github.com/jbloomlab/seqneut-pipeline](https://github.com/jbloomlab/seqneut-pipeline)\\nfor the computer code and underlying numerical data.\\n\\nSee [here](https://github.com/jbloomlab/seqneut-pipeline/graphs/contributors) for a\\nlist of all contributors to the pipeline.\\n\\x94\\x8c\\x0fviral_libraries\\x94}\\x94\\x8c\\x14pdmH1N1_lib2023_loes\\x94\\x8c-data/viral_libraries/pdmH1N1_lib2023_loes.csv\\x94s\\x8c\\x17viral_strain_plot_order\\x94\\x8c data/viral_strain_plot_order.csv\\x94\\x8c\\x12neut_standard_sets\\x94}\\x94\\x8c\\x08loes2023\\x94\\x8c3data/neut_standard_sets/loes2023_neut_standards.csv\\x94s\\x8c\\x1eillumina_barcode_parser_params\\x94}\\x94(\\x8c\\x08upstream\\x94\\x8c\\x1fCTCCCTACAATGTCGGATTTGTATTTAATAG\\x94\\x8c\\ndownstream\\x94\\x8c\\x00\\x94\\x8c\\x04minq\\x94K\\x14\\x8c\\x11upstream_mismatch\\x94K\\x04\\x8c\\x0ebc_orientation\\x94\\x8c\\x02R2\\x94u\\x8c#default_process_plate_qc_thresholds\\x94}\\x94(h\\xdaK\\xfah\\xdbG?tz\\xe1G\\xae\\x14{h\\xdc}\\x94(h\\xdeK\\x03h\\xdfK\\x02uh\\xe0}\\x94(h\\xe2K\\x03h\\xe3K\\x02uh\\xe4K\\x04h\\xe5G?@bM\\xd2\\xf1\\xa9\\xfch\\xe6G?tz\\xe1G\\xae\\x14{h\\xe7M^\\x01h\\xe8K\\x05h\\xe9K\\x04h\\xeaK\\x08u\\x8c\\x17default_curvefit_params\\x94}\\x94(h\\xf4K\\x01h\\xf5\\x89h\\xf6K\\x00u\\x8c\\x06plates\\x94}\\x94(\\x8c\\x06plate2\\x94}\\x94(\\x8c\\x04date\\x94\\x8c\\x08datetime\\x94\\x8c\\x04date\\x94\\x93\\x94C\\x04\\x07\\xe7\\x08\\x01\\x94\\x85\\x94R\\x94\\x8c\\rviral_library\\x94\\x8c\\x14pdmH1N1_lib2023_loes\\x94\\x8c\\x11neut_standard_set\\x94\\x8c\\x08loes2023\\x94\\x8c\\x0bsamples_csv\\x94\\x8c\\x1edata/plates/plate2_samples.csv\\x94\\x8c\\rqc_thresholds\\x94}\\x94(h\\xdaK\\xfah\\xdbG?tz\\xe1G\\xae\\x14{h\\xdc}\\x94(h\\xdeK\\x03h\\xdfK\\x02uh\\xe0}\\x94(h\\xe2K\\x03h\\xe3K\\x02uh\\xe4K\\x04h\\xe5G?@bM\\xd2\\xf1\\xa9\\xfch\\xe6G?tz\\xe1G\\xae\\x14{h\\xe7M^\\x01h\\xe8K\\x05h\\xe9K\\x04h\\xeaK\\x08u\\x8c\\x10barcodes_to_drop\\x94]\\x94\\x8c\\x10GGTCCATCTCAGATCG\\x94a\\x8c\\rwells_to_drop\\x94]\\x94(\\x8c\\x02C1\\x94\\x8c\\x03D12\\x94e\\x8c\\x0fcurvefit_params\\x94}\\x94(h\\xf4K\\x01h\\xf5\\x89h\\xf6K\\x00uuj{\\x01\\x00\\x00}\\x94(h\\xd0j\\xf6\\x01\\x00\\x00C\\x04\\x07\\xe7\\t\\x1a\\x94\\x85\\x94R\\x94h\\xd2h\\xd3h\\xd4h\\xd5h\\xd6h\\xd7h\\xd8}\\x94(h\\xdaK\\xfah\\xdbG?tz\\xe1G\\xae\\x14{h\\xdc}\\x94(h\\xdeK\\x03h\\xdfK\\x02uh\\xe0}\\x94(h\\xe2K\\x03h\\xe3K\\x02uh\\xe4K\\x04h\\xe5G?@bM\\xd2\\xf1\\xa9\\xfch\\xe6G?tz\\xe1G\\xae\\x14{h\\xe7M^\\x01h\\xe8K\\x05h\\xe9K\\x04h\\xeaK\\x08uh\\xeb]\\x94h\\xedah\\xee]\\x94(h\\xf0h\\xf1eh\\xf2}\\x94(h\\xf4K\\x01h\\xf5\\x89h\\xf6K\\x00uuu\\x8c\\x1aserum_titers_qc_thresholds\\x94}\\x94(\\x8c\\x14min_frac_infectivity\\x94G?\\xe6ffffff\\x8c\\x0emin_replicates\\x94K\\x02\\x8c\\x1bmax_fold_change_from_median\\x94K\\x03u\\x8c\\x1aserum_titers_qc_exclusions\\x94}\\x94(\\x8c\\x06M099d0\\x94}\\x94(\\x8c\\x16A/Bangladesh/8002/2021\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x12A/Brisbane/02/2018\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x13A/Norway/25089/2022\\x94}\\x94\\x8c\\x12replicates_to_drop\\x94]\\x94\\x8c\\x18plate11-CGGATAAAAATGATAT\\x94as\\x8c\\x14A/Wisconsin/588/2019\\x94}\\x94\\x8c\\x12replicates_to_drop\\x94]\\x94\\x8c\\x18plate11-AGTCCTATCCTCAAAT\\x94as\\x8c\\x19A/SouthAfrica/R16462/2021\\x94}\\x94\\x8c\\x12replicates_to_drop\\x94]\\x94\\x8c\\x17plate2-CTAGCAGATTGTATAA\\x94asu\\x8c\\x07M099d30\\x94}\\x94(\\x8c\\x13A/Chester/5355/2022\\x94}\\x94\\x8c\\x12replicates_to_drop\\x94]\\x94\\x8c\\x18plate11-CCTCAAAATAACAAGC\\x94as\\x8c\\x12A/Michigan/45/2015\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x12A/Paris/31196/2021\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88su\\x8c\\x07Y044d30\\x94}\\x94(\\x8c\\x16A/Bangladesh/8036/2021\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x12A/Brisbane/48/2022\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x1aA/India-PUN-NIV328484/2021\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x0eA/Perth/1/2022\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88s\\x8c\\x14A/Washington/23/2020\\x94}\\x94\\x8c\\tignore_qc\\x94\\x88suuu\\x8c\\x04rule\\x94\\x8c\\rprocess_plate\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8ct/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline/notebooks\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/fh/fast/bloom_j/computational_notebooks/jbloom/2023/flu_seqneut_DRIVE_2021-22_repeat_vax/seqneut-pipeline/test_example');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "199f1499-8e9e-4a01-a41c-e7937c94995c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.167565Z",
     "iopub.status.busy": "2023-12-28T14:34:21.167305Z",
     "iopub.status.idle": "2023-12-28T14:34:21.184288Z",
     "shell.execute_reply": "2023-12-28T14:34:21.183134Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.167538Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing plate='plate11'\n",
      "\n",
      "Plate has 36 samples (wells)\n",
      "\n",
      "Using the following QC thresholds:\n",
      "avg_barcode_counts_per_well: 250\n",
      "barcode_frac_consistency: 4\n",
      "max_frac_infectivity: 8\n",
      "min_dilutions_per_serum_replicate: 4\n",
      "min_neut_standard_barcode_frac: 0.005\n",
      "min_neut_standard_count: 350\n",
      "min_neut_standard_frac_per_well: 0.005\n",
      "min_no_serum_viral_barcode_count: 5\n",
      "min_viral_barcode_frac: 0.0005\n",
      "neut_standard_barcode_consistency: {max_barcodes: 2, max_fold_change: 3}\n",
      "no_serum_barcode_consistency: {max_fold_change: 3, max_wells: 2}\n"
     ]
    }
   ],
   "source": [
    "count_csvs = snakemake.input.count_csvs\n",
    "fate_csvs = snakemake.input.fate_csvs\n",
    "viral_library_csv = snakemake.input.viral_library_csv\n",
    "neut_standard_set_csv = snakemake.input.neut_standard_set_csv\n",
    "frac_infectivity_csv = snakemake.output.frac_infectivity_csv\n",
    "samples = snakemake.params.samples\n",
    "plate = snakemake.wildcards.plate\n",
    "plate_params = snakemake.params.plate_params\n",
    "qc_thresholds = snakemake.params.plate_params[\"qc_thresholds\"]\n",
    "\n",
    "print(f\"Processing {plate=}\")\n",
    "\n",
    "samples_df = pd.DataFrame(plate_params[\"samples\"])\n",
    "print(f\"\\nPlate has {len(samples)} samples (wells)\")\n",
    "assert all(\n",
    "    (len(samples_df) == samples_df[c].nunique())\n",
    "    for c in [\"well\", \"sample\", \"sample_noplate\"]\n",
    ")\n",
    "assert (\n",
    "    len(samples_df) == len(samples_df.groupby([\"serum_replicate\", \"dilution_factor\"]))\n",
    ")\n",
    "assert len(samples) == len(count_csvs) == len(fate_csvs) == len(samples_df)\n",
    "\n",
    "print(\"\\nUsing the following QC thresholds:\")\n",
    "y = yaml.YAML(typ=\"safe\").dump(qc_thresholds, stream=sys.stdout)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbf4bda2-e16f-4fc3-adbb-b5d3a7bfe2e5",
   "metadata": {},
   "source": [
    "Set up dictionary where we will keep track of wells, barcodes, well-barcodes, and serum-replicates that are dropped:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6aab62a4-f32f-49cc-bb4e-0e96d5a2d1ad",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.185773Z",
     "iopub.status.busy": "2023-12-28T14:34:21.185391Z",
     "iopub.status.idle": "2023-12-28T14:34:21.189905Z",
     "shell.execute_reply": "2023-12-28T14:34:21.188912Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.185742Z"
    }
   },
   "outputs": [],
   "source": [
    "qc_drops = collections.defaultdict(dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc6f5cd-de3d-4824-a421-7b460090d72e",
   "metadata": {},
   "source": [
    "## Statistics on barcode-parsing for each sample\n",
    "Make interactive chart of the \"fates\" of the sequencing reads parsed for each sample on the plate.\n",
    "\n",
    "If most sequencing reads are not \"valid barcodes\", this could potentially indicate some problem in the sequencing or barcode set you are parsing.\n",
    "\n",
    "Potential fates are:\n",
    " - *valid barcode*: barcode that matches a known virus or neutralization standard, we hope most reads are this.\n",
    " - *invalid barcode*: a barcode with proper flanking sequences, but does not match a known virus or neutralization standard. If you  have a lot of reads of this type, it is probably a good idea to look at the invalid barcode CSVs (in the `./results/barcode_invalid/` subdirectory created by the pipeline) to see what these invalid barcodes are.\n",
    " - *unparseable barcode*: could not parse a barcode from this read as there was not a sequence of the correct length with the appropriate flanking sequence.\n",
    " - *low quality barcode*: low-quality or `N` nucleotides in barcode, could indicate problem with sequencing.\n",
    " - *failed chastity filter*: reads that failed the Illumina chastity filter, if these are reported in the FASTQ (they may not be).\n",
    "\n",
    "Also, if the number of reads per sample is very uneven, that could indicate that you did not do a good job of balancing the different samples in the Illumina sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9aa103c7-3240-4dac-8532-f7298a5fe009",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.191330Z",
     "iopub.status.busy": "2023-12-28T14:34:21.190964Z",
     "iopub.status.idle": "2023-12-28T14:34:21.387580Z",
     "shell.execute_reply": "2023-12-28T14:34:21.386463Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.191299Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-76d4945348c248a3afb6a38687f5b12a.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-76d4945348c248a3afb6a38687f5b12a.vega-embed details,\n",
       "  #altair-viz-76d4945348c248a3afb6a38687f5b12a.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-76d4945348c248a3afb6a38687f5b12a\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-76d4945348c248a3afb6a38687f5b12a\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-76d4945348c248a3afb6a38687f5b12a\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.16.3?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.16.3\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}, \"axis\": {\"grid\": false}}, \"data\": {\"name\": \"data-65c88dd4b5091024fc8b6f005f84637e\"}, \"mark\": {\"type\": \"bar\", \"height\": {\"band\": 0.85}}, \"encoding\": {\"color\": {\"field\": \"fate\", \"sort\": [\"valid barcode\", \"unparseable barcode\", \"low quality barcode\", \"invalid barcode\"], \"type\": \"nominal\"}, \"order\": {\"field\": \"fate\", \"sort\": \"descending\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"fate\", \"type\": \"nominal\"}, {\"field\": \"count\", \"type\": \"quantitative\"}, {\"field\": \"well\", \"type\": \"nominal\"}, {\"field\": \"serum_replicate\", \"type\": \"nominal\"}, {\"field\": \"sample_well\", \"type\": \"nominal\"}, {\"field\": \"dilution_factor\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"count\", \"scale\": {\"nice\": false, \"padding\": 3}, \"type\": \"quantitative\"}, \"y\": {\"field\": \"sample_well\", \"sort\": [\"M099d0_20 (C2)\", \"M099d0_20 (C2)\", \"M099d0_20 (C2)\", \"M099d0_20 (C2)\", \"M099d0_60 (C3)\", \"M099d0_60 (C3)\", \"M099d0_60 (C3)\", \"M099d0_60 (C3)\", \"M099d0_180 (C4)\", \"M099d0_180 (C4)\", \"M099d0_180 (C4)\", \"M099d0_180 (C4)\", \"M099d0_540 (C5)\", \"M099d0_540 (C5)\", \"M099d0_540 (C5)\", \"M099d0_540 (C5)\", \"M099d0_1620 (C6)\", \"M099d0_1620 (C6)\", \"M099d0_1620 (C6)\", \"M099d0_1620 (C6)\", \"M099d0_4860 (C7)\", \"M099d0_4860 (C7)\", \"M099d0_4860 (C7)\", \"M099d0_4860 (C7)\", \"M099d0_14580 (C8)\", \"M099d0_14580 (C8)\", \"M099d0_14580 (C8)\", \"M099d0_14580 (C8)\", \"M099d0_43740 (C9)\", \"M099d0_43740 (C9)\", \"M099d0_43740 (C9)\", \"M099d0_43740 (C9)\", \"M099d0_131220 (C10)\", \"M099d0_131220 (C10)\", \"M099d0_131220 (C10)\", \"M099d0_131220 (C10)\", \"M099d0_393660 (C11)\", \"M099d0_393660 (C11)\", \"M099d0_393660 (C11)\", \"M099d0_393660 (C11)\", \"M099d30_20 (B2)\", \"M099d30_20 (B2)\", \"M099d30_20 (B2)\", \"M099d30_20 (B2)\", \"M099d30_60 (B3)\", \"M099d30_60 (B3)\", \"M099d30_60 (B3)\", \"M099d30_60 (B3)\", \"M099d30_180 (B4)\", \"M099d30_180 (B4)\", \"M099d30_180 (B4)\", \"M099d30_180 (B4)\", \"M099d30_540 (B5)\", \"M099d30_540 (B5)\", \"M099d30_540 (B5)\", \"M099d30_540 (B5)\", \"M099d30_1620 (B6)\", \"M099d30_1620 (B6)\", \"M099d30_1620 (B6)\", \"M099d30_1620 (B6)\", \"M099d30_4860 (B7)\", \"M099d30_4860 (B7)\", \"M099d30_4860 (B7)\", \"M099d30_4860 (B7)\", \"M099d30_14580 (B8)\", \"M099d30_14580 (B8)\", \"M099d30_14580 (B8)\", \"M099d30_14580 (B8)\", \"M099d30_43740 (B9)\", \"M099d30_43740 (B9)\", \"M099d30_43740 (B9)\", \"M099d30_43740 (B9)\", \"M099d30_131220 (B10)\", \"M099d30_131220 (B10)\", \"M099d30_131220 (B10)\", \"M099d30_131220 (B10)\", \"M099d30_393660 (B11)\", \"M099d30_393660 (B11)\", \"M099d30_393660 (B11)\", \"M099d30_393660 (B11)\", \"Y044d30_20 (D2)\", \"Y044d30_20 (D2)\", \"Y044d30_20 (D2)\", \"Y044d30_20 (D2)\", \"Y044d30_60 (D3)\", \"Y044d30_60 (D3)\", \"Y044d30_60 (D3)\", \"Y044d30_60 (D3)\", \"Y044d30_180 (D4)\", \"Y044d30_180 (D4)\", \"Y044d30_180 (D4)\", \"Y044d30_180 (D4)\", \"Y044d30_540 (D5)\", \"Y044d30_540 (D5)\", \"Y044d30_540 (D5)\", \"Y044d30_540 (D5)\", \"Y044d30_1620 (D6)\", \"Y044d30_1620 (D6)\", \"Y044d30_1620 (D6)\", \"Y044d30_1620 (D6)\", \"Y044d30_4860 (D7)\", \"Y044d30_4860 (D7)\", \"Y044d30_4860 (D7)\", \"Y044d30_4860 (D7)\", \"Y044d30_14580 (D8)\", \"Y044d30_14580 (D8)\", \"Y044d30_14580 (D8)\", \"Y044d30_14580 (D8)\", \"Y044d30_43740 (D9)\", \"Y044d30_43740 (D9)\", \"Y044d30_43740 (D9)\", \"Y044d30_43740 (D9)\", \"Y044d30_131220 (D10)\", \"Y044d30_131220 (D10)\", \"Y044d30_131220 (D10)\", \"Y044d30_131220 (D10)\", \"Y044d30_393660 (D11)\", \"Y044d30_393660 (D11)\", \"Y044d30_393660 (D11)\", \"Y044d30_393660 (D11)\", \"none-10 (B1)\", \"none-10 (B1)\", \"none-10 (B1)\", \"none-10 (B1)\", \"none-11 (C1)\", \"none-11 (C1)\", \"none-11 (C1)\", \"none-11 (C1)\", \"none-12 (D1)\", \"none-12 (D1)\", \"none-12 (D1)\", \"none-12 (D1)\", \"none-2 (B12)\", \"none-2 (B12)\", \"none-2 (B12)\", \"none-2 (B12)\", \"none-3 (C12)\", \"none-3 (C12)\", \"none-3 (C12)\", \"none-3 (C12)\", \"none-4 (D12)\", \"none-4 (D12)\", \"none-4 (D12)\", \"none-4 (D12)\"], \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 10}, \"params\": [{\"name\": \"param_2\", \"select\": {\"type\": \"point\", \"fields\": [\"serum_replicate\"]}, \"bind\": {\"input\": \"select\", \"options\": [null, \"M099d0\", \"M099d30\", \"Y044d30\", \"none-10\", \"none-11\", \"none-12\", \"none-2\", \"none-3\", \"none-4\"], \"labels\": [\"all\", \"M099d0\", \"M099d30\", \"Y044d30\", \"none-10\", \"none-11\", \"none-12\", \"none-2\", \"none-3\", \"none-4\"], \"name\": \"serum\"}}], \"title\": \"Barcode parsing for plate11\", \"transform\": [{\"filter\": {\"param\": \"param_2\"}}], \"width\": 200, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.16.3.json\", \"datasets\": {\"data-65c88dd4b5091024fc8b6f005f84637e\": [{\"fate\": \"valid barcode\", \"count\": 38187, \"well\": \"B1\", \"serum_replicate\": \"none-10\", \"sample_well\": \"none-10 (B1)\", \"dilution_factor\": null}, {\"fate\": \"low quality barcode\", \"count\": 8366, \"well\": \"B1\", \"serum_replicate\": \"none-10\", \"sample_well\": \"none-10 (B1)\", \"dilution_factor\": null}, {\"fate\": \"invalid barcode\", \"count\": 2777, \"well\": \"B1\", \"serum_replicate\": \"none-10\", \"sample_well\": \"none-10 (B1)\", \"dilution_factor\": null}, {\"fate\": \"unparseable barcode\", \"count\": 670, \"well\": \"B1\", \"serum_replicate\": \"none-10\", \"sample_well\": \"none-10 (B1)\", \"dilution_factor\": null}, {\"fate\": \"valid barcode\", \"count\": 40829, \"well\": \"B2\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_20 (B2)\", \"dilution_factor\": 20.0}, {\"fate\": \"low quality barcode\", \"count\": 7810, \"well\": \"B2\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_20 (B2)\", \"dilution_factor\": 20.0}, {\"fate\": \"unparseable barcode\", \"count\": 1025, \"well\": \"B2\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_20 (B2)\", \"dilution_factor\": 20.0}, {\"fate\": \"invalid barcode\", \"count\": 336, \"well\": \"B2\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_20 (B2)\", \"dilution_factor\": 20.0}, {\"fate\": \"valid barcode\", \"count\": 40956, \"well\": \"B3\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_60 (B3)\", \"dilution_factor\": 60.0}, {\"fate\": \"low quality barcode\", \"count\": 7654, \"well\": \"B3\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_60 (B3)\", \"dilution_factor\": 60.0}, {\"fate\": \"unparseable barcode\", \"count\": 990, \"well\": \"B3\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_60 (B3)\", \"dilution_factor\": 60.0}, {\"fate\": \"invalid barcode\", \"count\": 400, \"well\": \"B3\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_60 (B3)\", \"dilution_factor\": 60.0}, {\"fate\": \"valid barcode\", \"count\": 41677, \"well\": \"B4\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_180 (B4)\", \"dilution_factor\": 180.0}, {\"fate\": \"low quality barcode\", \"count\": 6998, \"well\": \"B4\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_180 (B4)\", \"dilution_factor\": 180.0}, {\"fate\": \"unparseable barcode\", \"count\": 951, \"well\": \"B4\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_180 (B4)\", \"dilution_factor\": 180.0}, {\"fate\": \"invalid barcode\", \"count\": 374, \"well\": \"B4\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_180 (B4)\", \"dilution_factor\": 180.0}, {\"fate\": \"valid barcode\", \"count\": 35899, \"well\": \"B5\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_540 (B5)\", \"dilution_factor\": 540.0}, {\"fate\": \"low quality barcode\", \"count\": 7508, \"well\": \"B5\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_540 (B5)\", \"dilution_factor\": 540.0}, {\"fate\": \"invalid barcode\", \"count\": 5692, \"well\": \"B5\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_540 (B5)\", \"dilution_factor\": 540.0}, {\"fate\": \"unparseable barcode\", \"count\": 901, \"well\": \"B5\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_540 (B5)\", \"dilution_factor\": 540.0}, {\"fate\": \"valid barcode\", \"count\": 34970, \"well\": \"B6\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_1620 (B6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"low quality barcode\", \"count\": 8235, \"well\": \"B6\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_1620 (B6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"invalid barcode\", \"count\": 6157, \"well\": \"B6\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_1620 (B6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"unparseable barcode\", \"count\": 638, \"well\": \"B6\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_1620 (B6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"valid barcode\", \"count\": 38189, \"well\": \"B7\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_4860 (B7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"low quality barcode\", \"count\": 8315, \"well\": \"B7\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_4860 (B7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"invalid barcode\", \"count\": 2843, \"well\": \"B7\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_4860 (B7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"unparseable barcode\", \"count\": 653, \"well\": \"B7\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_4860 (B7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"valid barcode\", \"count\": 38478, \"well\": \"B8\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_14580 (B8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"low quality barcode\", \"count\": 8134, \"well\": \"B8\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_14580 (B8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"invalid barcode\", \"count\": 2721, \"well\": \"B8\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_14580 (B8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"unparseable barcode\", \"count\": 667, \"well\": \"B8\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_14580 (B8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"valid barcode\", \"count\": 37189, \"well\": \"B9\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_43740 (B9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"low quality barcode\", \"count\": 8155, \"well\": \"B9\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_43740 (B9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"invalid barcode\", \"count\": 4001, \"well\": \"B9\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_43740 (B9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"unparseable barcode\", \"count\": 655, \"well\": \"B9\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_43740 (B9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"valid barcode\", \"count\": 37998, \"well\": \"B10\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_131220 (B10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"low quality barcode\", \"count\": 8381, \"well\": \"B10\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_131220 (B10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"invalid barcode\", \"count\": 2963, \"well\": \"B10\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_131220 (B10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"unparseable barcode\", \"count\": 658, \"well\": \"B10\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_131220 (B10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"valid barcode\", \"count\": 38758, \"well\": \"B11\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_393660 (B11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"low quality barcode\", \"count\": 8459, \"well\": \"B11\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_393660 (B11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"invalid barcode\", \"count\": 2114, \"well\": \"B11\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_393660 (B11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"unparseable barcode\", \"count\": 669, \"well\": \"B11\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_393660 (B11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"valid barcode\", \"count\": 37756, \"well\": \"B12\", \"serum_replicate\": \"none-2\", \"sample_well\": \"none-2 (B12)\", \"dilution_factor\": null}, {\"fate\": \"low quality barcode\", \"count\": 8493, \"well\": \"B12\", \"serum_replicate\": \"none-2\", \"sample_well\": \"none-2 (B12)\", \"dilution_factor\": null}, {\"fate\": \"invalid barcode\", \"count\": 3104, \"well\": \"B12\", \"serum_replicate\": \"none-2\", \"sample_well\": \"none-2 (B12)\", \"dilution_factor\": null}, {\"fate\": \"unparseable barcode\", \"count\": 647, \"well\": \"B12\", \"serum_replicate\": \"none-2\", \"sample_well\": \"none-2 (B12)\", \"dilution_factor\": null}, {\"fate\": \"valid barcode\", \"count\": 38228, \"well\": \"C1\", \"serum_replicate\": \"none-11\", \"sample_well\": \"none-11 (C1)\", \"dilution_factor\": null}, {\"fate\": \"low quality barcode\", \"count\": 8050, \"well\": \"C1\", \"serum_replicate\": \"none-11\", \"sample_well\": \"none-11 (C1)\", \"dilution_factor\": null}, {\"fate\": \"invalid barcode\", \"count\": 3080, \"well\": \"C1\", \"serum_replicate\": \"none-11\", \"sample_well\": \"none-11 (C1)\", \"dilution_factor\": null}, {\"fate\": \"unparseable barcode\", \"count\": 642, \"well\": \"C1\", \"serum_replicate\": \"none-11\", \"sample_well\": \"none-11 (C1)\", \"dilution_factor\": null}, {\"fate\": \"valid barcode\", \"count\": 40558, \"well\": \"C2\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_20 (C2)\", \"dilution_factor\": 20.0}, {\"fate\": \"low quality barcode\", \"count\": 8078, \"well\": \"C2\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_20 (C2)\", \"dilution_factor\": 20.0}, {\"fate\": \"unparseable barcode\", \"count\": 1025, \"well\": \"C2\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_20 (C2)\", \"dilution_factor\": 20.0}, {\"fate\": \"invalid barcode\", \"count\": 339, \"well\": \"C2\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_20 (C2)\", \"dilution_factor\": 20.0}, {\"fate\": \"valid barcode\", \"count\": 40701, \"well\": \"C3\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_60 (C3)\", \"dilution_factor\": 60.0}, {\"fate\": \"low quality barcode\", \"count\": 7831, \"well\": \"C3\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_60 (C3)\", \"dilution_factor\": 60.0}, {\"fate\": \"unparseable barcode\", \"count\": 1044, \"well\": \"C3\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_60 (C3)\", \"dilution_factor\": 60.0}, {\"fate\": \"invalid barcode\", \"count\": 424, \"well\": \"C3\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_60 (C3)\", \"dilution_factor\": 60.0}, {\"fate\": \"valid barcode\", \"count\": 39092, \"well\": \"C4\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_180 (C4)\", \"dilution_factor\": 180.0}, {\"fate\": \"low quality barcode\", \"count\": 8051, \"well\": \"C4\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_180 (C4)\", \"dilution_factor\": 180.0}, {\"fate\": \"invalid barcode\", \"count\": 2105, \"well\": \"C4\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_180 (C4)\", \"dilution_factor\": 180.0}, {\"fate\": \"unparseable barcode\", \"count\": 752, \"well\": \"C4\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_180 (C4)\", \"dilution_factor\": 180.0}, {\"fate\": \"valid barcode\", \"count\": 38873, \"well\": \"C5\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_540 (C5)\", \"dilution_factor\": 540.0}, {\"fate\": \"low quality barcode\", \"count\": 8567, \"well\": \"C5\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_540 (C5)\", \"dilution_factor\": 540.0}, {\"fate\": \"invalid barcode\", \"count\": 1446, \"well\": \"C5\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_540 (C5)\", \"dilution_factor\": 540.0}, {\"fate\": \"unparseable barcode\", \"count\": 1114, \"well\": \"C5\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_540 (C5)\", \"dilution_factor\": 540.0}, {\"fate\": \"valid barcode\", \"count\": 37799, \"well\": \"C6\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_1620 (C6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"low quality barcode\", \"count\": 8432, \"well\": \"C6\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_1620 (C6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"invalid barcode\", \"count\": 3139, \"well\": \"C6\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_1620 (C6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"unparseable barcode\", \"count\": 630, \"well\": \"C6\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_1620 (C6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"valid barcode\", \"count\": 37649, \"well\": \"C7\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_4860 (C7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"low quality barcode\", \"count\": 7910, \"well\": \"C7\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_4860 (C7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"invalid barcode\", \"count\": 3749, \"well\": \"C7\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_4860 (C7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"unparseable barcode\", \"count\": 692, \"well\": \"C7\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_4860 (C7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"valid barcode\", \"count\": 38288, \"well\": \"C8\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_14580 (C8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"low quality barcode\", \"count\": 7750, \"well\": \"C8\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_14580 (C8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"invalid barcode\", \"count\": 3257, \"well\": \"C8\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_14580 (C8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"unparseable barcode\", \"count\": 705, \"well\": \"C8\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_14580 (C8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"valid barcode\", \"count\": 37504, \"well\": \"C9\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_43740 (C9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"low quality barcode\", \"count\": 8347, \"well\": \"C9\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_43740 (C9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"invalid barcode\", \"count\": 3520, \"well\": \"C9\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_43740 (C9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"unparseable barcode\", \"count\": 629, \"well\": \"C9\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_43740 (C9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"valid barcode\", \"count\": 37793, \"well\": \"C10\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_131220 (C10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"low quality barcode\", \"count\": 8315, \"well\": \"C10\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_131220 (C10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"invalid barcode\", \"count\": 3092, \"well\": \"C10\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_131220 (C10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"unparseable barcode\", \"count\": 800, \"well\": \"C10\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_131220 (C10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"valid barcode\", \"count\": 38716, \"well\": \"C11\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_393660 (C11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"low quality barcode\", \"count\": 8322, \"well\": \"C11\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_393660 (C11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"invalid barcode\", \"count\": 2224, \"well\": \"C11\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_393660 (C11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"unparseable barcode\", \"count\": 738, \"well\": \"C11\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_393660 (C11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"valid barcode\", \"count\": 37828, \"well\": \"C12\", \"serum_replicate\": \"none-3\", \"sample_well\": \"none-3 (C12)\", \"dilution_factor\": null}, {\"fate\": \"low quality barcode\", \"count\": 8061, \"well\": \"C12\", \"serum_replicate\": \"none-3\", \"sample_well\": \"none-3 (C12)\", \"dilution_factor\": null}, {\"fate\": \"invalid barcode\", \"count\": 3386, \"well\": \"C12\", \"serum_replicate\": \"none-3\", \"sample_well\": \"none-3 (C12)\", \"dilution_factor\": null}, {\"fate\": \"unparseable barcode\", \"count\": 725, \"well\": \"C12\", \"serum_replicate\": \"none-3\", \"sample_well\": \"none-3 (C12)\", \"dilution_factor\": null}, {\"fate\": \"valid barcode\", \"count\": 37911, \"well\": \"D1\", \"serum_replicate\": \"none-12\", \"sample_well\": \"none-12 (D1)\", \"dilution_factor\": null}, {\"fate\": \"low quality barcode\", \"count\": 8456, \"well\": \"D1\", \"serum_replicate\": \"none-12\", \"sample_well\": \"none-12 (D1)\", \"dilution_factor\": null}, {\"fate\": \"invalid barcode\", \"count\": 3015, \"well\": \"D1\", \"serum_replicate\": \"none-12\", \"sample_well\": \"none-12 (D1)\", \"dilution_factor\": null}, {\"fate\": \"unparseable barcode\", \"count\": 618, \"well\": \"D1\", \"serum_replicate\": \"none-12\", \"sample_well\": \"none-12 (D1)\", \"dilution_factor\": null}, {\"fate\": \"valid barcode\", \"count\": 41659, \"well\": \"D2\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_20 (D2)\", \"dilution_factor\": 20.0}, {\"fate\": \"low quality barcode\", \"count\": 7050, \"well\": \"D2\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_20 (D2)\", \"dilution_factor\": 20.0}, {\"fate\": \"unparseable barcode\", \"count\": 965, \"well\": \"D2\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_20 (D2)\", \"dilution_factor\": 20.0}, {\"fate\": \"invalid barcode\", \"count\": 326, \"well\": \"D2\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_20 (D2)\", \"dilution_factor\": 20.0}, {\"fate\": \"valid barcode\", \"count\": 40396, \"well\": \"D3\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_60 (D3)\", \"dilution_factor\": 60.0}, {\"fate\": \"low quality barcode\", \"count\": 8242, \"well\": \"D3\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_60 (D3)\", \"dilution_factor\": 60.0}, {\"fate\": \"unparseable barcode\", \"count\": 1023, \"well\": \"D3\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_60 (D3)\", \"dilution_factor\": 60.0}, {\"fate\": \"invalid barcode\", \"count\": 339, \"well\": \"D3\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_60 (D3)\", \"dilution_factor\": 60.0}, {\"fate\": \"valid barcode\", \"count\": 40797, \"well\": \"D4\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_180 (D4)\", \"dilution_factor\": 180.0}, {\"fate\": \"low quality barcode\", \"count\": 7722, \"well\": \"D4\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_180 (D4)\", \"dilution_factor\": 180.0}, {\"fate\": \"unparseable barcode\", \"count\": 1009, \"well\": \"D4\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_180 (D4)\", \"dilution_factor\": 180.0}, {\"fate\": \"invalid barcode\", \"count\": 472, \"well\": \"D4\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_180 (D4)\", \"dilution_factor\": 180.0}, {\"fate\": \"valid barcode\", \"count\": 34315, \"well\": \"D5\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_540 (D5)\", \"dilution_factor\": 540.0}, {\"fate\": \"low quality barcode\", \"count\": 8600, \"well\": \"D5\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_540 (D5)\", \"dilution_factor\": 540.0}, {\"fate\": \"invalid barcode\", \"count\": 6267, \"well\": \"D5\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_540 (D5)\", \"dilution_factor\": 540.0}, {\"fate\": \"unparseable barcode\", \"count\": 818, \"well\": \"D5\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_540 (D5)\", \"dilution_factor\": 540.0}, {\"fate\": \"valid barcode\", \"count\": 29509, \"well\": \"D6\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_1620 (D6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"invalid barcode\", \"count\": 11603, \"well\": \"D6\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_1620 (D6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"low quality barcode\", \"count\": 8215, \"well\": \"D6\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_1620 (D6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"unparseable barcode\", \"count\": 673, \"well\": \"D6\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_1620 (D6)\", \"dilution_factor\": 1620.0}, {\"fate\": \"valid barcode\", \"count\": 35386, \"well\": \"D7\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_4860 (D7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"low quality barcode\", \"count\": 8531, \"well\": \"D7\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_4860 (D7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"invalid barcode\", \"count\": 5425, \"well\": \"D7\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_4860 (D7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"unparseable barcode\", \"count\": 658, \"well\": \"D7\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_4860 (D7)\", \"dilution_factor\": 4860.0}, {\"fate\": \"valid barcode\", \"count\": 35997, \"well\": \"D8\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_14580 (D8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"low quality barcode\", \"count\": 8364, \"well\": \"D8\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_14580 (D8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"invalid barcode\", \"count\": 5061, \"well\": \"D8\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_14580 (D8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"unparseable barcode\", \"count\": 578, \"well\": \"D8\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_14580 (D8)\", \"dilution_factor\": 14580.0}, {\"fate\": \"valid barcode\", \"count\": 38052, \"well\": \"D9\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_43740 (D9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"low quality barcode\", \"count\": 7421, \"well\": \"D9\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_43740 (D9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"invalid barcode\", \"count\": 3913, \"well\": \"D9\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_43740 (D9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"unparseable barcode\", \"count\": 614, \"well\": \"D9\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_43740 (D9)\", \"dilution_factor\": 43740.0}, {\"fate\": \"valid barcode\", \"count\": 37412, \"well\": \"D10\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_131220 (D10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"low quality barcode\", \"count\": 8167, \"well\": \"D10\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_131220 (D10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"invalid barcode\", \"count\": 3754, \"well\": \"D10\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_131220 (D10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"unparseable barcode\", \"count\": 667, \"well\": \"D10\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_131220 (D10)\", \"dilution_factor\": 131220.0}, {\"fate\": \"valid barcode\", \"count\": 37989, \"well\": \"D11\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_393660 (D11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"low quality barcode\", \"count\": 7911, \"well\": \"D11\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_393660 (D11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"invalid barcode\", \"count\": 3493, \"well\": \"D11\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_393660 (D11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"unparseable barcode\", \"count\": 607, \"well\": \"D11\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_393660 (D11)\", \"dilution_factor\": 393660.0}, {\"fate\": \"valid barcode\", \"count\": 38369, \"well\": \"D12\", \"serum_replicate\": \"none-4\", \"sample_well\": \"none-4 (D12)\", \"dilution_factor\": null}, {\"fate\": \"low quality barcode\", \"count\": 8225, \"well\": \"D12\", \"serum_replicate\": \"none-4\", \"sample_well\": \"none-4 (D12)\", \"dilution_factor\": null}, {\"fate\": \"invalid barcode\", \"count\": 2787, \"well\": \"D12\", \"serum_replicate\": \"none-4\", \"sample_well\": \"none-4 (D12)\", \"dilution_factor\": null}, {\"fate\": \"unparseable barcode\", \"count\": 619, \"well\": \"D12\", \"serum_replicate\": \"none-4\", \"sample_well\": \"none-4 (D12)\", \"dilution_factor\": null}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fates = (\n",
    "    pd.concat([pd.read_csv(f).assign(sample=s) for f, s in zip(fate_csvs, samples)])\n",
    "    .merge(samples_df, validate=\"many_to_one\", on=\"sample\")\n",
    "    .assign(\n",
    "        fate_counts=lambda x: x.groupby(\"fate\")[\"count\"].transform(\"sum\"),\n",
    "        sample_well=lambda x: x[\"sample_noplate\"] + \" (\" + x[\"well\"] + \")\",\n",
    "    )\n",
    "    .query(\"fate_counts > 0\")[  # only keep fates with at least one count\n",
    "        [\"fate\", \"count\", \"well\", \"serum_replicate\", \"sample_well\", \"dilution_factor\"]\n",
    "    ]\n",
    ")\n",
    "\n",
    "assert len(fates) == len(fates.drop_duplicates())\n",
    "\n",
    "serum_replicates = sorted(fates[\"serum_replicate\"].unique())\n",
    "sample_wells = list(\n",
    "    fates.sort_values([\"serum_replicate\", \"dilution_factor\"])[\"sample_well\"]\n",
    ")\n",
    "            \n",
    "\n",
    "serum_selection = alt.selection_point(\n",
    "    fields=[\"serum_replicate\"],\n",
    "    bind=alt.binding_select(\n",
    "        options=[None] + serum_replicates,\n",
    "        labels=[\"all\"] + serum_replicates,\n",
    "        name=\"serum\",\n",
    "    ),\n",
    ")\n",
    "\n",
    "fates_chart = (\n",
    "    alt.Chart(fates)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\"count\", scale=alt.Scale(nice=False, padding=3)),\n",
    "        alt.Y(\n",
    "            \"sample_well\",\n",
    "            title=None,\n",
    "            sort=sample_wells,\n",
    "        ),\n",
    "        alt.Color(\"fate\", sort=sorted(fates[\"fate\"].unique(), reverse=True)),\n",
    "        alt.Order(\"fate\", sort=\"descending\"),\n",
    "        tooltip=fates.columns.tolist(),\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=200,\n",
    "        title=f\"Barcode parsing for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "fates_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a810fb9a-189d-48e5-b866-6ef0f2e0dd5f",
   "metadata": {},
   "source": [
    "## Counts per barcode\n",
    "\n",
    "First get the counts per barcode:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ccea55e6-d280-4e8e-998a-1a86930c6529",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.389652Z",
     "iopub.status.busy": "2023-12-28T14:34:21.389390Z",
     "iopub.status.idle": "2023-12-28T14:34:21.488285Z",
     "shell.execute_reply": "2023-12-28T14:34:21.487126Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.389625Z"
    }
   },
   "outputs": [],
   "source": [
    "# get barcode counts\n",
    "counts = (\n",
    "    pd.concat([pd.read_csv(c).assign(sample=s) for c, s in zip(count_csvs, samples)])\n",
    "    .merge(samples_df, validate=\"many_to_one\", on=\"sample\")\n",
    "    .drop(columns=[\"replicate\", \"plate\", \"fastq\"])\n",
    "    .assign(sample_well=lambda x: x[\"sample_noplate\"] + \" (\" + x[\"well\"] + \")\")\n",
    ")\n",
    "\n",
    "# classify barcodes as viral or neut standard\n",
    "barcode_class = pd.concat(\n",
    "    [\n",
    "        pd.read_csv(viral_library_csv)[[\"barcode\", \"strain\"]].assign(\n",
    "            neut_standard=False,\n",
    "        ),\n",
    "        pd.read_csv(neut_standard_set_csv)[[\"barcode\"]].assign(\n",
    "            neut_standard=True,\n",
    "            strain=pd.NA,\n",
    "        ),\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "# merge counts and classification of barcodes\n",
    "assert set(counts[\"barcode\"]) == set(barcode_class[\"barcode\"])\n",
    "counts = counts.merge(barcode_class, on=\"barcode\", validate=\"many_to_one\")\n",
    "assert set(sample_wells) == set(counts[\"sample_well\"])\n",
    "assert set(serum_replicates) == set(counts[\"serum_replicate\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ddb2544-a1a3-460f-9d81-fa061b152071",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-05T19:26:10.449100Z",
     "iopub.status.busy": "2023-12-05T19:26:10.448471Z",
     "iopub.status.idle": "2023-12-05T19:26:10.468023Z",
     "shell.execute_reply": "2023-12-05T19:26:10.467117Z",
     "shell.execute_reply.started": "2023-12-05T19:26:10.449041Z"
    }
   },
   "source": [
    "Plot average counts per barcode.\n",
    "If a sample has inadequate barcode counts, it may not have good enough statistics for accurate analysis, and a QC-threshold is applied:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "68da99a3-e1e4-4b03-a7c2-79406e6d7170",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-28T14:34:21.489757Z",
     "iopub.status.busy": "2023-12-28T14:34:21.489493Z",
     "iopub.status.idle": "2023-12-28T14:34:21.605400Z",
     "shell.execute_reply": "2023-12-28T14:34:21.604189Z",
     "shell.execute_reply.started": "2023-12-28T14:34:21.489729Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-a937091021864478a366cb2dea8462c5.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-a937091021864478a366cb2dea8462c5.vega-embed details,\n",
       "  #altair-viz-a937091021864478a366cb2dea8462c5.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-a937091021864478a366cb2dea8462c5\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-a937091021864478a366cb2dea8462c5\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-a937091021864478a366cb2dea8462c5\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.16.3?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.16.3\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}, \"axis\": {\"grid\": false}}, \"data\": {\"name\": \"data-332c6d5ff89cbe54867a5c99174e90ea\"}, \"mark\": {\"type\": \"bar\", \"height\": {\"band\": 0.85}}, \"encoding\": {\"color\": {\"field\": \"fails_qc\", \"legend\": {\"titleLimit\": 500}, \"title\": \"fails qc_thresholds['avg_barcode_counts_per_well']=250\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"well\", \"type\": \"nominal\"}, {\"field\": \"serum_replicate\", \"type\": \"nominal\"}, {\"field\": \"sample_well\", \"type\": \"nominal\"}, {\"field\": \"avg_count\", \"format\": \".3g\", \"type\": \"quantitative\"}, {\"field\": \"fails_qc\", \"type\": \"nominal\"}], \"x\": {\"field\": \"avg_count\", \"scale\": {\"nice\": false, \"padding\": 3}, \"title\": \"average barcode counts per well\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"sample_well\", \"sort\": [\"M099d0_20 (C2)\", \"M099d0_20 (C2)\", \"M099d0_20 (C2)\", \"M099d0_20 (C2)\", \"M099d0_60 (C3)\", \"M099d0_60 (C3)\", \"M099d0_60 (C3)\", \"M099d0_60 (C3)\", \"M099d0_180 (C4)\", \"M099d0_180 (C4)\", \"M099d0_180 (C4)\", \"M099d0_180 (C4)\", \"M099d0_540 (C5)\", \"M099d0_540 (C5)\", \"M099d0_540 (C5)\", \"M099d0_540 (C5)\", \"M099d0_1620 (C6)\", \"M099d0_1620 (C6)\", \"M099d0_1620 (C6)\", \"M099d0_1620 (C6)\", \"M099d0_4860 (C7)\", \"M099d0_4860 (C7)\", \"M099d0_4860 (C7)\", \"M099d0_4860 (C7)\", \"M099d0_14580 (C8)\", \"M099d0_14580 (C8)\", \"M099d0_14580 (C8)\", \"M099d0_14580 (C8)\", \"M099d0_43740 (C9)\", \"M099d0_43740 (C9)\", \"M099d0_43740 (C9)\", \"M099d0_43740 (C9)\", \"M099d0_131220 (C10)\", \"M099d0_131220 (C10)\", \"M099d0_131220 (C10)\", \"M099d0_131220 (C10)\", \"M099d0_393660 (C11)\", \"M099d0_393660 (C11)\", \"M099d0_393660 (C11)\", \"M099d0_393660 (C11)\", \"M099d30_20 (B2)\", \"M099d30_20 (B2)\", \"M099d30_20 (B2)\", \"M099d30_20 (B2)\", \"M099d30_60 (B3)\", \"M099d30_60 (B3)\", \"M099d30_60 (B3)\", \"M099d30_60 (B3)\", \"M099d30_180 (B4)\", \"M099d30_180 (B4)\", \"M099d30_180 (B4)\", \"M099d30_180 (B4)\", \"M099d30_540 (B5)\", \"M099d30_540 (B5)\", \"M099d30_540 (B5)\", \"M099d30_540 (B5)\", \"M099d30_1620 (B6)\", \"M099d30_1620 (B6)\", \"M099d30_1620 (B6)\", \"M099d30_1620 (B6)\", \"M099d30_4860 (B7)\", \"M099d30_4860 (B7)\", \"M099d30_4860 (B7)\", \"M099d30_4860 (B7)\", \"M099d30_14580 (B8)\", \"M099d30_14580 (B8)\", \"M099d30_14580 (B8)\", \"M099d30_14580 (B8)\", \"M099d30_43740 (B9)\", \"M099d30_43740 (B9)\", \"M099d30_43740 (B9)\", \"M099d30_43740 (B9)\", \"M099d30_131220 (B10)\", \"M099d30_131220 (B10)\", \"M099d30_131220 (B10)\", \"M099d30_131220 (B10)\", \"M099d30_393660 (B11)\", \"M099d30_393660 (B11)\", \"M099d30_393660 (B11)\", \"M099d30_393660 (B11)\", \"Y044d30_20 (D2)\", \"Y044d30_20 (D2)\", \"Y044d30_20 (D2)\", \"Y044d30_20 (D2)\", \"Y044d30_60 (D3)\", \"Y044d30_60 (D3)\", \"Y044d30_60 (D3)\", \"Y044d30_60 (D3)\", \"Y044d30_180 (D4)\", \"Y044d30_180 (D4)\", \"Y044d30_180 (D4)\", \"Y044d30_180 (D4)\", \"Y044d30_540 (D5)\", \"Y044d30_540 (D5)\", \"Y044d30_540 (D5)\", \"Y044d30_540 (D5)\", \"Y044d30_1620 (D6)\", \"Y044d30_1620 (D6)\", \"Y044d30_1620 (D6)\", \"Y044d30_1620 (D6)\", \"Y044d30_4860 (D7)\", \"Y044d30_4860 (D7)\", \"Y044d30_4860 (D7)\", \"Y044d30_4860 (D7)\", \"Y044d30_14580 (D8)\", \"Y044d30_14580 (D8)\", \"Y044d30_14580 (D8)\", \"Y044d30_14580 (D8)\", \"Y044d30_43740 (D9)\", \"Y044d30_43740 (D9)\", \"Y044d30_43740 (D9)\", \"Y044d30_43740 (D9)\", \"Y044d30_131220 (D10)\", \"Y044d30_131220 (D10)\", \"Y044d30_131220 (D10)\", \"Y044d30_131220 (D10)\", \"Y044d30_393660 (D11)\", \"Y044d30_393660 (D11)\", \"Y044d30_393660 (D11)\", \"Y044d30_393660 (D11)\", \"none-10 (B1)\", \"none-10 (B1)\", \"none-10 (B1)\", \"none-10 (B1)\", \"none-11 (C1)\", \"none-11 (C1)\", \"none-11 (C1)\", \"none-11 (C1)\", \"none-12 (D1)\", \"none-12 (D1)\", \"none-12 (D1)\", \"none-12 (D1)\", \"none-2 (B12)\", \"none-2 (B12)\", \"none-2 (B12)\", \"none-2 (B12)\", \"none-3 (C12)\", \"none-3 (C12)\", \"none-3 (C12)\", \"none-3 (C12)\", \"none-4 (D12)\", \"none-4 (D12)\", \"none-4 (D12)\", \"none-4 (D12)\"], \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 10}, \"params\": [{\"name\": \"param_2\", \"select\": {\"type\": \"point\", \"fields\": [\"serum_replicate\"]}, \"bind\": {\"input\": \"select\", \"options\": [null, \"M099d0\", \"M099d30\", \"Y044d30\", \"none-10\", \"none-11\", \"none-12\", \"none-2\", \"none-3\", \"none-4\"], \"labels\": [\"all\", \"M099d0\", \"M099d30\", \"Y044d30\", \"none-10\", \"none-11\", \"none-12\", \"none-2\", \"none-3\", \"none-4\"], \"name\": \"serum\"}}], \"title\": \"Average barcode counts for plate11\", \"transform\": [{\"filter\": {\"param\": \"param_2\"}}], \"width\": 250, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.16.3.json\", \"datasets\": {\"data-332c6d5ff89cbe54867a5c99174e90ea\": [{\"well\": \"B1\", \"serum_replicate\": \"none-10\", \"sample_well\": \"none-10 (B1)\", \"avg_count\": 318.225, \"fails_qc\": false}, {\"well\": \"B10\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_131220 (B10)\", \"avg_count\": 316.65, \"fails_qc\": false}, {\"well\": \"B11\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_393660 (B11)\", \"avg_count\": 322.98333333333335, \"fails_qc\": false}, {\"well\": \"B12\", \"serum_replicate\": \"none-2\", \"sample_well\": \"none-2 (B12)\", \"avg_count\": 314.6333333333333, \"fails_qc\": false}, {\"well\": \"B2\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_20 (B2)\", \"avg_count\": 340.2416666666667, \"fails_qc\": false}, {\"well\": \"B3\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_60 (B3)\", \"avg_count\": 341.3, \"fails_qc\": false}, {\"well\": \"B4\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_180 (B4)\", \"avg_count\": 347.30833333333334, \"fails_qc\": false}, {\"well\": \"B5\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_540 (B5)\", \"avg_count\": 299.15833333333336, \"fails_qc\": false}, {\"well\": \"B6\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_1620 (B6)\", \"avg_count\": 291.4166666666667, \"fails_qc\": false}, {\"well\": \"B7\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_4860 (B7)\", \"avg_count\": 318.2416666666667, \"fails_qc\": false}, {\"well\": \"B8\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_14580 (B8)\", \"avg_count\": 320.65, \"fails_qc\": false}, {\"well\": \"B9\", \"serum_replicate\": \"M099d30\", \"sample_well\": \"M099d30_43740 (B9)\", \"avg_count\": 309.90833333333336, \"fails_qc\": false}, {\"well\": \"C1\", \"serum_replicate\": \"none-11\", \"sample_well\": \"none-11 (C1)\", \"avg_count\": 318.56666666666666, \"fails_qc\": false}, {\"well\": \"C10\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_131220 (C10)\", \"avg_count\": 314.94166666666666, \"fails_qc\": false}, {\"well\": \"C11\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_393660 (C11)\", \"avg_count\": 322.6333333333333, \"fails_qc\": false}, {\"well\": \"C12\", \"serum_replicate\": \"none-3\", \"sample_well\": \"none-3 (C12)\", \"avg_count\": 315.23333333333335, \"fails_qc\": false}, {\"well\": \"C2\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_20 (C2)\", \"avg_count\": 337.98333333333335, \"fails_qc\": false}, {\"well\": \"C3\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_60 (C3)\", \"avg_count\": 339.175, \"fails_qc\": false}, {\"well\": \"C4\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_180 (C4)\", \"avg_count\": 325.76666666666665, \"fails_qc\": false}, {\"well\": \"C5\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_540 (C5)\", \"avg_count\": 323.94166666666666, \"fails_qc\": false}, {\"well\": \"C6\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_1620 (C6)\", \"avg_count\": 314.9916666666667, \"fails_qc\": false}, {\"well\": \"C7\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_4860 (C7)\", \"avg_count\": 313.7416666666667, \"fails_qc\": false}, {\"well\": \"C8\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_14580 (C8)\", \"avg_count\": 319.06666666666666, \"fails_qc\": false}, {\"well\": \"C9\", \"serum_replicate\": \"M099d0\", \"sample_well\": \"M099d0_43740 (C9)\", \"avg_count\": 312.53333333333336, \"fails_qc\": false}, {\"well\": \"D1\", \"serum_replicate\": \"none-12\", \"sample_well\": \"none-12 (D1)\", \"avg_count\": 315.925, \"fails_qc\": false}, {\"well\": \"D10\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_131220 (D10)\", \"avg_count\": 311.76666666666665, \"fails_qc\": false}, {\"well\": \"D11\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_393660 (D11)\", \"avg_count\": 316.575, \"fails_qc\": false}, {\"well\": \"D12\", \"serum_replicate\": \"none-4\", \"sample_well\": \"none-4 (D12)\", \"avg_count\": 319.7416666666667, \"fails_qc\": false}, {\"well\": \"D2\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_20 (D2)\", \"avg_count\": 347.15833333333336, \"fails_qc\": false}, {\"well\": \"D3\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_60 (D3)\", \"avg_count\": 336.6333333333333, \"fails_qc\": false}, {\"well\": \"D4\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_180 (D4)\", \"avg_count\": 339.975, \"fails_qc\": false}, {\"well\": \"D5\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_540 (D5)\", \"avg_count\": 285.9583333333333, \"fails_qc\": false}, {\"well\": \"D6\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_1620 (D6)\", \"avg_count\": 245.90833333333333, \"fails_qc\": true}, {\"well\": \"D7\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_4860 (D7)\", \"avg_count\": 294.8833333333333, \"fails_qc\": false}, {\"well\": \"D8\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_14580 (D8)\", \"avg_count\": 299.975, \"fails_qc\": false}, {\"well\": \"D9\", \"serum_replicate\": \"Y044d30\", \"sample_well\": \"Y044d30_43740 (D9)\", \"avg_count\": 317.1, \"fails_qc\": false}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Dropping 1 wells for failing qc_thresholds['avg_barcode_counts_per_well']=250:\n",
      "  ['D6']\n",
      "4320\n",
      "4200\n"
     ]
    }
   ],
   "source": [
    "avg_barcode_counts = (\n",
    "    counts.groupby(\n",
    "        [\"well\", \"serum_replicate\", \"sample_well\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(avg_count=pd.NamedAgg(\"count\", \"mean\"))\n",
    "    .assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"avg_count\"] < qc_thresholds['avg_barcode_counts_per_well']\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "avg_barcode_counts_chart = (\n",
    "    alt.Chart(avg_barcode_counts)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"avg_count\",\n",
    "            title=\"average barcode counts per well\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\"sample_well\", title=None, sort=sample_wells),\n",
    "        alt.Color(\n",
    "            \"fails_qc\",\n",
    "            title=f\"fails {qc_thresholds['avg_barcode_counts_per_well']=}\",\n",
    "            legend=alt.Legend(titleLimit=500),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if avg_barcode_counts[c].dtype == float else c\n",
    "            for c in avg_barcode_counts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Average barcode counts for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "display(avg_barcode_counts_chart)\n",
    "\n",
    "# drop wells failing QC\n",
    "avg_barcode_counts_per_well_drops = list(avg_barcode_counts.query(\"fails_qc\")[\"well\"])\n",
    "print(\n",
    "    f\"\\nDropping {len(avg_barcode_counts_per_well_drops)} wells for failing \"\n",
    "    f\"{qc_thresholds['avg_barcode_counts_per_well']=}:\\n  \"\n",
    "    + str(avg_barcode_counts_per_well_drops)\n",
    ")\n",
    "qc_drops[\"wells\"].update(\n",
    "    {w: \"avg_barcode_counts_per_well\" for w in avg_barcode_counts_per_well_drops}\n",
    ")\n",
    "counts = counts[~counts[\"well\"].isin(qc_drops[\"wells\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589a755d-b46f-46ea-a84c-2dd784fa8672",
   "metadata": {},
   "source": [
    "## Fraction of counts from neutralization standard\n",
    "Determine the fraction of counts from the neutralization standard in each sample, and make sure this fraction passess the QC threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a71183-4670-40c1-88c1-237bb536803e",
   "metadata": {},
   "outputs": [],
   "source": [
    "neut_standard_fracs = (\n",
    "    counts.assign(\n",
    "        neut_standard_count=lambda x: x[\"count\"] * x[\"neut_standard\"].astype(int)\n",
    "    )\n",
    "    .groupby(\n",
    "        [\"well\", \"serum_replicate\", \"sample_noplate\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(\n",
    "        total_count=pd.NamedAgg(\"count\", \"sum\"),\n",
    "        neut_standard_count=pd.NamedAgg(\"neut_standard_count\", \"sum\"),\n",
    "    )\n",
    "    .assign(\n",
    "        neut_standard_frac=lambda x: x[\"neut_standard_count\"] / x[\"total_count\"],\n",
    "    )\n",
    ")\n",
    "\n",
    "neut_standard_fracs_chart = (\n",
    "    alt.Chart(neut_standard_fracs)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"neut_standard_frac\",\n",
    "            title=\"fraction of counts from neutralization standard\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"sample_noplate\",\n",
    "            title=None,\n",
    "            sort=samples_noplate,\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if neut_standard_fracs[c].dtype == float else c\n",
    "            for c in neut_standard_fracs.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Neutralization-standard fractions for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(neut_standard_fracs_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e990a5d-075e-4e81-971c-14483f1dd7ea",
   "metadata": {},
   "source": [
    "## Consistency and minimum fractions for barcodes\n",
    "We examine the fraction of counts attributable to each barcode. We do this splitting the data two ways:\n",
    "\n",
    " 1. Looking at all viral (but not neut-standard) barcodes only for the no-serum samples.\n",
    "\n",
    " 2. Looking at just the neut-standard barcodes for all samples.\n",
    "\n",
    "The reasons is that if the experiment is set up perfectly, these fractions should be the same across all samples for each barcode.\n",
    "(We do not expect viral barcodes to have consistent fractions across no-serum samples as they will be neutralized differently depending on strain).\n",
    "\n",
    "We plot these fractions in interactive plots (you can mouseover points and zoom) so you can identify barcodes that fail the expected consistency QC thresholds.\n",
    "\n",
    "We also make sure the barcodes meet specified QC minimum thresholds for all samples, and flag any that do not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fecec428-ad2f-4877-bcd4-a399fc7b7d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_selection = alt.selection_point(fields=[\"barcode\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "# look at all samples for neut standard barcodes, or no-serum samples for all barcodes\n",
    "for is_neut_standard, df in counts.groupby(\"neut_standard\"):\n",
    "    # process data frame\n",
    "    if not is_neut_standard:\n",
    "        df = df.query(\"serum == 'none'\")\n",
    "    df = df.assign(\n",
    "        sample_counts=lambda x: x.groupby(\"sample\")[\"count\"].transform(\"sum\"),\n",
    "        count_frac=lambda x: x[\"count\"] / x[\"sample_counts\"],\n",
    "        median_count_frac=lambda x: x.groupby(\"barcode\")[\"count_frac\"].transform(\n",
    "            \"median\"\n",
    "        ),\n",
    "        fold_change_from_median=lambda x: x[\"count_frac\"] / x[\"median_count_frac\"],\n",
    "    ).drop(\n",
    "        columns=(\n",
    "            [\"sample\", \"serum\", \"sample_counts\", \"neut_standard\", \"dilution_factor\", \"plate_replicate\"]\n",
    "            + ([\"strain\"] if is_neut_standard else [\"serum_replicate\"])\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    # make chart\n",
    "    evenness_chart = (\n",
    "        alt.Chart(df)\n",
    "        .add_params(barcode_selection)\n",
    "        .encode(\n",
    "            alt.X(\n",
    "                \"count_frac\",\n",
    "                title=(\n",
    "                    \"barcode's fraction of neut standard counts\"\n",
    "                    if is_neut_standard\n",
    "                    else \"barcode's fraction of non-neut standard counts\"\n",
    "                ),\n",
    "                scale=alt.Scale(nice=False, padding=5),\n",
    "            ),\n",
    "            alt.Y(\"sample_noplate\", title=None, sort=samples_noplate),\n",
    "            alt.Fill(\"barcode\", legend=None),\n",
    "            strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "            size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3g\") if df[c].dtype == float else c\n",
    "                for c in df.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_circle(fillOpacity=0.6, stroke=\"black\", strokeOpacity=1)\n",
    "        .properties(\n",
    "            height=alt.Step(10),\n",
    "            width=300,\n",
    "            title=(\n",
    "                f\"{plate} all samples, neut-standard barcodes\"\n",
    "                if is_neut_standard\n",
    "                else f\"{plate} no-serum samples, all barcodes\"\n",
    "            ),\n",
    "        )\n",
    "        .configure_axis(grid=False)\n",
    "        .configure_legend(titleLimit=1000)\n",
    "        .interactive()\n",
    "    )\n",
    "\n",
    "    if is_neut_standard:\n",
    "        evenness_chart = evenness_chart.add_params(serum_selection).transform_filter(\n",
    "            serum_selection\n",
    "        )\n",
    "        print(f\"\\n\\n{'=' * 89}\\nAnalyzing neut-standard barcodes from all samples\\n\")\n",
    "    else:\n",
    "        print(f\"\\n\\n{'=' * 89}\\nAnalyzing all barcodes from no-serum samples\\n\")\n",
    "\n",
    "    display(evenness_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743e4e41-aef9-4980-8b59-55d7b0a686fd",
   "metadata": {},
   "source": [
    "## Compute fraction infectivity\n",
    "\n",
    "The fraction infectivity for viral barcode $v_b$ in sample $s$ is computed as:\n",
    "$$\n",
    "F_{v_b,s} = \\frac{c_{v_b,s} / \\left(\\sum_{n_b} c_{n_b,s}\\right)}{{\\rm median}_{s_0}\\left[ c_{v_b,s_0} / \\left(\\sum_{n_b} c_{n_b,s_0}\\right)\\right]}\n",
    "$$\n",
    "where\n",
    " - $c_{v_b,s}$ is the counts of viral barcode $v_b$ in sample $s$.\n",
    " - $\\sum_{n_b} c_{n_b,s}$ is the sum of the counts for all neutralization standard barcodes $n_b$ for sample $s$.\n",
    " - $c_{v_b,s_0}$ is the counts of viral barcode $v_b$ in no-serum sample $s_0$.\n",
    " - $\\sum_{n_b} c_{n_b,s_0}$ is the sum of the counts for all neutralization standard barcodes $n_b$ for no-serum sample $s_0$.\n",
    " - ${\\rm median}_{s_0}\\left[ c_{v_b,s_0} / \\left(\\sum_{n_b} c_{n_b,s_0}\\right)\\right]$ is the median taken across all no-serum samples of the counts of viral barcode $v_b$ versus the total counts for all neutralization standard barcodes.\n",
    "\n",
    "First, compute the total neutralization-standard counts for each sample.\n",
    "Plot these, and make sure they meet the QC threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ef1afa4-5772-4596-a68d-b98e080ba2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "neut_standard_counts = (\n",
    "    counts.query(\"neut_standard\")\n",
    "    .groupby(\n",
    "        [\"well\", \"serum_replicate\", \"sample_noplate\", \"dilution_factor\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(neut_standard_count=pd.NamedAgg(\"count\", \"sum\"))\n",
    ")\n",
    "\n",
    "neut_standard_counts_chart = (\n",
    "    alt.Chart(neut_standard_counts)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"neut_standard_count\",\n",
    "            title=\"counts from neutralization standard\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\"sample_noplate\", title=None, sort=samples_noplate),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\")\n",
    "            if neut_standard_counts[c].dtype == float\n",
    "            else c\n",
    "            for c in neut_standard_counts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Neutralization-standard counts for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(neut_standard_counts_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9baafb57-e601-46d4-bed4-4bd5eb61552b",
   "metadata": {},
   "source": [
    "Compute and plot the no-serum sample viral barcode counts and check if they pass the QC filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d060ad12-b159-49e0-8ff4-94d175da5bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_serum_counts = (\n",
    "    counts.query(\"serum == 'none'\")\n",
    "    .query(\"not neut_standard\")\n",
    "    .merge(neut_standard_counts, validate=\"many_to_one\")[\n",
    "        [\"barcode\", \"strain\", \"well\", \"sample_noplate\", \"count\", \"neut_standard_count\"]\n",
    "    ]\n",
    ")\n",
    "\n",
    "strains = sorted(no_serum_counts[\"strain\"].unique())\n",
    "strain_selection_dropdown = alt.selection_point(\n",
    "    fields=[\"strain\"],\n",
    "    bind=alt.binding_select(\n",
    "        options=[None] + strains, labels=[\"all\"] + strains, name=\"virus strain\",\n",
    "    ),\n",
    ")\n",
    "\n",
    "# make chart\n",
    "no_serum_counts_chart = (\n",
    "    alt.Chart(no_serum_counts)\n",
    "    .add_params(barcode_selection, strain_selection_dropdown)\n",
    "    .transform_filter(strain_selection_dropdown)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"count\", title=\"viral barcode count\", scale=alt.Scale(nice=False, padding=5)\n",
    "        ),\n",
    "        alt.Y(\"sample_noplate\", title=None),\n",
    "        alt.Fill(\"barcode\", legend=None),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "        size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "        tooltip=no_serum_counts.columns.tolist(),\n",
    "    )\n",
    "    .mark_circle(fillOpacity=0.6, stroke=\"black\", strokeOpacity=1)\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=400,\n",
    "        title=f\"{plate} viral barcode counts in no-serum samples\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    "    .interactive()\n",
    ")\n",
    "\n",
    "display(no_serum_counts_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdffe0e7-9864-4c4f-851d-5604106ff63a",
   "metadata": {},
   "source": [
    "Compute and plot the median ratio of viral barcode count to neut standard counts across no-serum samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b15a0c-9d54-48f4-86e3-f38b2015a5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "median_no_serum_ratio = (\n",
    "    no_serum_counts.assign(ratio=lambda x: x[\"count\"] / x[\"neut_standard_count\"])\n",
    "    .groupby([\"barcode\", \"strain\"], as_index=False)\n",
    "    .aggregate(median_no_serum_ratio=pd.NamedAgg(\"ratio\", \"median\"))\n",
    ")\n",
    "\n",
    "strain_selection = alt.selection_point(fields=[\"strain\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "median_no_serum_ratio_chart = (\n",
    "    alt.Chart(median_no_serum_ratio)\n",
    "    .add_params(strain_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"median_no_serum_ratio\",\n",
    "            title=\"median ratio of counts\",\n",
    "            scale=alt.Scale(nice=False, padding=5),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"barcode\",\n",
    "            sort=alt.SortField(\"median_no_serum_ratio\", order=\"descending\"),\n",
    "            axis=alt.Axis(labelFontSize=5),\n",
    "        ),\n",
    "        color=alt.condition(strain_selection, alt.value(\"orange\"), alt.value(\"gray\")),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\")\n",
    "            if median_no_serum_ratio[c].dtype == float\n",
    "            else c\n",
    "            for c in median_no_serum_ratio.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(5),\n",
    "        width=250,\n",
    "        title=f\"{plate} no-serum median ratio viral barcode to neut-standard barcode\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(median_no_serum_ratio_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb39004a-3684-4501-afd0-da259c85b131",
   "metadata": {},
   "source": [
    "Compute the actual fraction infectivities:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d787193c-3c50-427d-a0ec-165b446d8e73",
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_infectivity = (\n",
    "    counts.query(\"not neut_standard\")\n",
    "    .query(\"serum != 'none'\")\n",
    "    .merge(median_no_serum_ratio, validate=\"many_to_one\")\n",
    "    .merge(neut_standard_counts, validate=\"many_to_one\")\n",
    "    .assign(\n",
    "        frac_infectivity=lambda x: (\n",
    "            (x[\"count\"] / x[\"neut_standard_count\"]) / x[\"median_no_serum_ratio\"]\n",
    "        ),\n",
    "        concentration=lambda x: 1 / x[\"dilution_factor\"],\n",
    "        plate_barcode=lambda x: x[\"plate_replicate\"] + \"-\" + x[\"barcode\"],\n",
    "    )[\n",
    "        [\n",
    "            \"barcode\",\n",
    "            \"plate_barcode\",\n",
    "            \"strain\",\n",
    "            \"serum\",\n",
    "            \"serum_replicate\",\n",
    "            \"dilution_factor\",\n",
    "            \"concentration\",\n",
    "            \"frac_infectivity\",\n",
    "            \"well\",\n",
    "        ]\n",
    "    ]\n",
    ")\n",
    "\n",
    "assert len(\n",
    "    frac_infectivity.groupby([\"serum\", \"plate_barcode\", \"dilution_factor\"])\n",
    ") == len(frac_infectivity)\n",
    "assert frac_infectivity[\"dilution_factor\"].notnull().all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "465014e0-aced-4fb2-a503-620c1636c804",
   "metadata": {},
   "source": [
    "Plot the fraction infectivities:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982d871b-c140-4295-afe0-ea138bdafbaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_infectivity_chart_df = frac_infectivity[\n",
    "    [\"barcode\", \"strain\", \"serum_replicate\", \"dilution_factor\", \"frac_infectivity\", \"well\"]\n",
    "]\n",
    "\n",
    "frac_infectivity_chart = (\n",
    "    alt.Chart(frac_infectivity_chart_df)\n",
    "    .add_params(strain_selection_dropdown, barcode_selection)\n",
    "    .transform_filter(strain_selection_dropdown)\n",
    "    .encode(\n",
    "        alt.X(\"dilution_factor\", title=\"dilution factor\", scale=alt.Scale(nice=False, padding=3, type=\"log\")),\n",
    "        alt.Y(\"frac_infectivity\", title=\"fraction infectivity\", scale=alt.Scale(nice=False, padding=3)),\n",
    "        alt.Facet(\"serum_replicate\", title=None, columns=4, header=alt.Header(labelFontSize=13, labelFontStyle=\"bold\")),\n",
    "        alt.Detail(\"barcode\"),\n",
    "        color=alt.condition(barcode_selection, alt.value(\"black\"), alt.value(\"MediumBlue\")),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(3), alt.value(1)),\n",
    "        opacity=alt.condition(barcode_selection, alt.value(1), alt.value(0.25)),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if frac_infectivity_chart_df[c].dtype == float else c\n",
    "            for c in frac_infectivity_chart_df.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_line(point=True)\n",
    "    .properties(\n",
    "        height=250,\n",
    "        width=275,\n",
    "        title=f\"Fraction infectivities for {plate}\",\n",
    "    )\n",
    "    .interactive(bind_x=False)\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    "    .configure_point(size=45)\n",
    "    .resolve_scale(x=\"independent\", y=\"independent\")\n",
    ")\n",
    "\n",
    "display(frac_infectivity_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdb4eea-9822-4f27-89e6-6f11033bcb7f",
   "metadata": {},
   "source": [
    "Summarize key properties for each serum-replicate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e8998c6-b2de-43b8-9091-b65423257bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for prop, col, agg in [\n",
    "    (\"number of dilutions\", \"dilution_factor\", \"nunique\"),\n",
    "    (\"max frac infectivity\", \"frac_infectivity\", \"max\"),\n",
    "    (\"min frac infectivity\", \"frac_infectivity\", \"min\"),\n",
    "]:\n",
    "    prop_summary = (\n",
    "        frac_infectivity\n",
    "        .groupby([\"serum_replicate\", \"strain\", \"barcode\"], as_index=False)\n",
    "        .aggregate(**{prop: pd.NamedAgg(col, agg)})\n",
    "    )\n",
    "\n",
    "    prop_summary_chart = (\n",
    "        alt.Chart(prop_summary)\n",
    "        .add_params(barcode_selection)\n",
    "        .encode(\n",
    "            alt.X(prop, scale=alt.Scale(nice=False, padding=4)),\n",
    "            alt.Y(\"strain\", title=None),\n",
    "            alt.Column(\n",
    "                \"serum_replicate\",\n",
    "                title=None,\n",
    "                header=alt.Header(labelFontSize=12, labelFontStyle=\"bold\", labelPadding=0),\n",
    "            ),\n",
    "            strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "            size=alt.condition(barcode_selection, alt.value(55), alt.value(35)),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3g\") if prop_summary[c].dtype == float else c\n",
    "                for c in prop_summary.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_circle(stroke=\"black\", strokeOpacity=1, fillOpacity=0.45)\n",
    "        .properties(\n",
    "            height=alt.Step(10),\n",
    "            width=120,\n",
    "            title=alt.TitleParams(f\"{prop} for each serum-replicate\", dy=-2),\n",
    "        )\n",
    "    )\n",
    "\n",
    "    display(prop_summary_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8822240e-5846-48da-ad31-53adaf647910",
   "metadata": {},
   "source": [
    "## Fit neutralization curves\n",
    "First fit curves to all serum replicates, then we will filter out poor-fitting ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51174a6b-e023-4dc3-812a-25e54b8b73f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fits_nofilter = neutcurve.CurveFits(\n",
    "    frac_infectivity.rename(\n",
    "        columns={\n",
    "            \"frac_infectivity\": \"fraction infectivity\",\n",
    "            \"concentration\": \"serum concentration\",\n",
    "        }\n",
    "    ),\n",
    "    conc_col=\"serum concentration\",\n",
    "    fracinf_col=\"fraction infectivity\",\n",
    "    virus_col=\"strain\",\n",
    "    serum_col=\"serum\",\n",
    "    replicate_col=\"plate_barcode\",\n",
    ")\n",
    "\n",
    "_ = fits_nofilter.plotReplicates(\n",
    "    attempt_shared_legend=False,\n",
    "    legendfontsize=8,\n",
    "    titlesize=10,\n",
    "    ticksize=10,\n",
    "    ncol=6,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f0832bd-4632-4954-ae9d-ac65625ccc83",
   "metadata": {},
   "source": [
    "## Save results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69af6924-1dea-48fc-a9b0-e34ed411ef1d",
   "metadata": {},
   "source": [
    "Write fraction infectivities to file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70e5e6cc-f729-4509-8cb3-fabeefe38033",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"\\nWriting fraction infectivities to {frac_infectivity_csv}\")\n",
    "(\n",
    "    frac_infectivity[\n",
    "        [\"serum\", \"strain\", \"plate_barcode\", \"dilution_factor\", \"frac_infectivity\"]\n",
    "    ]\n",
    "    .sort_values([\"serum\", \"plate_barcode\", \"dilution_factor\"])\n",
    "    .to_csv(frac_infectivity_csv, index=False, float_format=\"%.5g\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28e34290-9cca-4ebc-9645-1e692ebb4848",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
