{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4eb38ea4-98fa-4da1-9071-2c65d38291ff",
   "metadata": {},
   "source": [
    "# Quality-control counts and use them to compute fraction infectivity\n",
    "This notebook is designed to be run using `snakemake`, and analyzes a plate of sequencing-based neutralization assays.\n",
    "\n",
    "## Setup\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "900fe170-1b4b-4646-b9ee-01d75a76f664",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import neutcurve\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc47332-2e29-4605-bf8b-463edb63cac6",
   "metadata": {},
   "source": [
    "Get the variables passed by `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "199f1499-8e9e-4a01-a41c-e7937c94995c",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_csvs = snakemake.input.count_csvs\n",
    "fate_csvs = snakemake.input.fate_csvs\n",
    "viral_library_csv = snakemake.input.viral_library_csv\n",
    "neut_standard_set_csv = snakemake.input.neut_standard_set_csv\n",
    "frac_infectivity_csv = snakemake.output.frac_infectivity_csv\n",
    "qc_failures_file = snakemake.output.qc_failures\n",
    "samples = snakemake.params.samples\n",
    "plate_params = snakemake.params.plate_params\n",
    "plate = snakemake.wildcards.plate\n",
    "\n",
    "samples_df = plate_params[\"samples\"]\n",
    "\n",
    "assert len(samples) == len(count_csvs) == len(fate_csvs) == len(samples_df)\n",
    "\n",
    "print(f\"Processing {plate=}\")\n",
    "\n",
    "# define set of QC failures for this plate\n",
    "qc_failures = set([])\n",
    "\n",
    "# get and print QC thresholds\n",
    "qc_thresholds = plate_params[\"process_counts_qc_thresholds\"]\n",
    "display(pd.Series(qc_thresholds))\n",
    "\n",
    "# will drop the following samples from `wells_to_drop`\n",
    "wells_to_drop = plate_params[\"wells_to_drop\"]\n",
    "if wells_to_drop:\n",
    "    if not set(wells_to_drop).issubset(samples_df[\"well\"]):\n",
    "        raise ValueError(f\"{wells_to_drop=} not all in `samples_df`\")\n",
    "    print(\"Dropping the following wells (samples):\")\n",
    "    display(samples_df.query(\"well in @wells_to_drop\"))\n",
    "    samples_df = samples_df.query(\"well not in @wells_to_drop\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc6f5cd-de3d-4824-a421-7b460090d72e",
   "metadata": {},
   "source": [
    "## Statistics on barcode-parsing for each sample\n",
    "Make interactive chart of the \"fates\" of the sequencing reads parsed for each sample on the plate.\n",
    "\n",
    "If most sequencing reads are not \"valid barcodes\", this could potentially indicate some problem in the sequencing or barcode set you are parsing.\n",
    "\n",
    "Potential fates are:\n",
    " - *valid barcode*: barcode that matches a known virus or neutralization standard, we hope most reads are this.\n",
    " - *invalid barcode*: a barcode with proper flanking sequences, but does not match a known virus or neutralization standard. If you  have a lot of reads of this type, it is probably a good idea to look at the invalid barcode CSVs (in the `./results/barcode_invalid/` subdirectory created by the pipeline) to see what these invalid barcodes are.\n",
    " - *unparseable barcode*: could not parse a barcode from this read as there was not a sequence of the correct length with the appropriate flanking sequence.\n",
    " - *low quality barcode*: low-quality or `N` nucleotides in barcode, could indicate problem with sequencing.\n",
    " - *failed chastity filter*: reads that failed the Illumina chastity filter, if these are reported in the FASTQ (they may not be).\n",
    "\n",
    "Also, if the number of reads per sample is very uneven, that could indicate that you did not do a good job of balancing the different samples in the Illumina sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa103c7-3240-4dac-8532-f7298a5fe009",
   "metadata": {},
   "outputs": [],
   "source": [
    "fates = (\n",
    "    pd.concat([pd.read_csv(f).assign(sample=s) for f, s in zip(fate_csvs, samples)])\n",
    "    .merge(samples_df, validate=\"many_to_one\", on=\"sample\")\n",
    "    .assign(fate_counts=lambda x: x.groupby(\"fate\")[\"count\"].transform(\"sum\"))\n",
    "    .query(\"fate_counts > 0\")  # only keep fates with at least one count\n",
    "    [[\"fate\", \"count\", \"well\", \"serum\", \"sample_noplate\", \"dilution_factor\"]]\n",
    ")\n",
    "\n",
    "assert len(fates) == len(fates.drop_duplicates())\n",
    "\n",
    "serum_selection = alt.selection_point(\n",
    "    fields=[\"serum\"],\n",
    "    bind=alt.binding_select(\n",
    "        options=[None] + sorted(fates[\"serum\"].unique().tolist()),\n",
    "        labels=[\"all\"] + [str(s) for s in sorted(fates[\"serum\"].unique())],\n",
    "        name=\"serum\",\n",
    "    )\n",
    ")   \n",
    "\n",
    "fates_chart = (\n",
    "    alt.Chart(fates)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\"count\", scale=alt.Scale(nice=False, padding=3)),\n",
    "        alt.Y(\n",
    "            \"sample_noplate\",\n",
    "            title=None,\n",
    "            sort=list(\n",
    "                fates.sort_values([\"serum\", \"dilution_factor\"])[\"sample_noplate\"]\n",
    "            ),\n",
    "        ),\n",
    "        alt.Color(\"fate\", sort=sorted(fates[\"fate\"].unique(), reverse=True)),\n",
    "        alt.Order(\"fate\", sort=\"descending\"),\n",
    "        tooltip=fates.columns.tolist(),\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=200,\n",
    "        title=f\"Barcode parsing for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "fates_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a810fb9a-189d-48e5-b866-6ef0f2e0dd5f",
   "metadata": {},
   "source": [
    "## Counts per barcode\n",
    "\n",
    "First get the counts per barcode and classification information on these barcodes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccea55e6-d280-4e8e-998a-1a86930c6529",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get barcode counts\n",
    "counts = (\n",
    "    pd.concat([pd.read_csv(c).assign(sample=s) for c, s in zip(count_csvs, samples)])\n",
    "    .merge(samples_df, validate=\"many_to_one\", on=\"sample\")\n",
    "    .drop(columns=[\"replicate\", \"plate\", \"fastq\"])\n",
    ")\n",
    "\n",
    "# get classification of barcodes as viral or neut standard\n",
    "barcode_class = pd.concat(\n",
    "    [\n",
    "        pd.read_csv(viral_library_csv)[[\"barcode\", \"strain\"]].assign(\n",
    "            neut_standard=False,\n",
    "        ),\n",
    "        pd.read_csv(neut_standard_set_csv)[[\"barcode\"]].assign(\n",
    "            neut_standard=True, strain=pd.NA,\n",
    "        ),\n",
    "    ],\n",
    "    ignore_index=True\n",
    ")\n",
    "\n",
    "# merge counts and classification of barcodes\n",
    "assert set(counts[\"barcode\"]) == set(barcode_class[\"barcode\"])\n",
    "counts = counts.merge(barcode_class, on=\"barcode\", validate=\"many_to_one\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b702fba-a8c2-4a83-b036-f1fc5d927298",
   "metadata": {},
   "source": [
    "Drop any barcodes that are specified to drop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f2adec2-557a-4575-b668-d12679ebe768",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes_to_drop = plate_params[\"barcodes_to_drop\"]\n",
    "\n",
    "if len(barcodes_to_drop):\n",
    "    print(\n",
    "        \"The following barcodes are specified to drop:\\n\\t\"\n",
    "        + \"\\n\\t\".join(barcodes_to_drop)\n",
    "    )\n",
    "    invalid_barcodes = set(barcodes_to_drop) - set(counts[\"barcode\"])\n",
    "    if invalid_barcodes:\n",
    "        raise ValueError(f\"Barcodes to drop do not exist: {invalid_barcodes}\")\n",
    "    counts = counts.query(\"barcode not in @barcodes_to_drop\")\n",
    "\n",
    "else:\n",
    "    print(f\"No barcodes specified to drop.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ddb2544-a1a3-460f-9d81-fa061b152071",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-05T19:26:10.449100Z",
     "iopub.status.busy": "2023-12-05T19:26:10.448471Z",
     "iopub.status.idle": "2023-12-05T19:26:10.468023Z",
     "shell.execute_reply": "2023-12-05T19:26:10.467117Z",
     "shell.execute_reply.started": "2023-12-05T19:26:10.449041Z"
    }
   },
   "source": [
    "Plot average counts per barcode, and make sure that these pass the QC threshold.\n",
    "If a sample has inadequate barcode counts, it may not have good enough statistics for accurate analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68da99a3-e1e4-4b03-a7c2-79406e6d7170",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_barcode_counts = (\n",
    "    counts\n",
    "    .groupby(\n",
    "        [\"well\", \"serum\", \"dilution_factor\", \"sample_noplate\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(avg_count=pd.NamedAgg(\"count\", \"mean\"))\n",
    "    .assign(passes_qc=lambda x: x[\"avg_count\"] >= qc_thresholds[\"avg_barcode_counts\"])\n",
    ")\n",
    "\n",
    "avg_barcode_counts_chart = (\n",
    "    alt.Chart(avg_barcode_counts)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"avg_count\",\n",
    "            title=\"average counts per barcode\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"sample_noplate\",\n",
    "            title=None,\n",
    "            sort=list(\n",
    "                avg_barcode_counts.sort_values([\"serum\", \"dilution_factor\"])[\"sample_noplate\"]\n",
    "            ),\n",
    "        ),\n",
    "        alt.Color(\n",
    "            \"passes_qc\",\n",
    "            title=f\"passes QC threshold {qc_thresholds['avg_barcode_counts']}\",\n",
    "            scale=alt.Scale(domain=[True, False]),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if avg_barcode_counts[c].dtype == float\n",
    "            else c\n",
    "            for c in avg_barcode_counts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Average barcode counts for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "display(avg_barcode_counts_chart)\n",
    "\n",
    "if not avg_barcode_counts[\"passes_qc\"].all():\n",
    "    qc_failures.add(\"avg_barcode_counts\")\n",
    "    print(f\"\\nThe following samples failed {qc_thresholds['avg_barcode_counts']=}\")\n",
    "    display(avg_barcode_counts.query(\"not passes_qc\").reset_index(drop=True))\n",
    "else:\n",
    "    print(f\"\\nAll samples passed {qc_thresholds['avg_barcode_counts']=}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589a755d-b46f-46ea-a84c-2dd784fa8672",
   "metadata": {},
   "source": [
    "## Fraction of counts from neutralization standard\n",
    "Determine the fraction of counts from the neutralization standard in each sample, and make sure this fraction passess the QC threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a71183-4670-40c1-88c1-237bb536803e",
   "metadata": {},
   "outputs": [],
   "source": [
    "neut_standard_fracs = (\n",
    "    counts\n",
    "    .assign(neut_standard_count=lambda x: x[\"count\"] * x[\"neut_standard\"].astype(int))\n",
    "    .groupby(\n",
    "        [\"well\", \"serum\", \"dilution_factor\", \"sample_noplate\"],\n",
    "        dropna=False,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(\n",
    "        total_count=pd.NamedAgg(\"count\", \"sum\"),\n",
    "        neut_standard_count=pd.NamedAgg(\"neut_standard_count\", \"sum\"),\n",
    "    )\n",
    "    .assign(\n",
    "        neut_standard_frac=lambda x: x[\"neut_standard_count\"] / x[\"total_count\"],\n",
    "        passes_qc=lambda x: (\n",
    "            (x[\"neut_standard_frac\"] >= qc_thresholds[\"min_neut_standard_frac\"])\n",
    "            & (\n",
    "                (x[\"serum\"] != \"none\")\n",
    "                | (x[\"neut_standard_frac\"] <= qc_thresholds[\"max_neut_standard_frac_no_serum\"])\n",
    "            )\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "neut_standard_qc_desc = (\n",
    "    f\"neut standard frac >= {qc_thresholds['min_neut_standard_frac']}, \"\n",
    "    + f\"<= {qc_thresholds['max_neut_standard_frac_no_serum']} for no-serum samples\"\n",
    ")\n",
    "\n",
    "neut_standard_fracs_chart = (\n",
    "    alt.Chart(neut_standard_fracs)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"neut_standard_frac\",\n",
    "            title=\"fraction of counts from neutralization standard\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"sample_noplate\",\n",
    "            title=None,\n",
    "            sort=list(\n",
    "                neut_standard_fracs.sort_values([\"serum\", \"dilution_factor\"])[\"sample_noplate\"]\n",
    "            ),\n",
    "        ),\n",
    "        alt.Color(\n",
    "            \"passes_qc\",\n",
    "            title=neut_standard_qc_desc,\n",
    "            scale=alt.Scale(domain=[True, False]),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if neut_standard_fracs[c].dtype == float\n",
    "            else c\n",
    "            for c in neut_standard_fracs.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Neutralization-standard fractions for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(neut_standard_fracs_chart)\n",
    "\n",
    "if not neut_standard_fracs[\"passes_qc\"].all():\n",
    "    qc_failures.add(\"min_neut_standard_frac or max_neut_standard_frac_no_serum\")\n",
    "    print(f\"\\nThe following samples failed {neut_standard_qc_desc}\")\n",
    "    display(neut_standard_fracs.query(\"not passes_qc\").reset_index(drop=True))\n",
    "else:\n",
    "    print(f\"\\nAll samples passed {neut_standard_qc_desc}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e990a5d-075e-4e81-971c-14483f1dd7ea",
   "metadata": {},
   "source": [
    "## Consistency and minimum fractions for barcodes\n",
    "We examine the fraction of counts attributable to each barcode. We do this splitting the data two ways:\n",
    "\n",
    " 1. Looking at all viral (but not neut-standard) barcodes only for the no-serum samples.\n",
    "\n",
    " 2. Looking at just the neut-standard barcodes for all samples.\n",
    "\n",
    "The reasons is that if the experiment is set up perfectly, these fractions should be the same across all samples for each barcode.\n",
    "(We do not expect viral barcodes to have consistent fractions across no-serum samples as they will be neutralized differently depending on strain).\n",
    "\n",
    "We plot these fractions in interactive plots (you can mouseover points and zoom) so you can identify barcodes that fail the expected consistency QC thresholds.\n",
    "\n",
    "We also make sure the barcodes meet specified QC minimum thresholds for all samples, and flag any that do not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fecec428-ad2f-4877-bcd4-a399fc7b7d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_selection = alt.selection_point(fields=[\"barcode\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "# look at all samples for neut standard barcodes, or no-serum samples for all barcodes\n",
    "for is_neut_standard, df in counts.groupby(\"neut_standard\"):\n",
    "\n",
    "    # process data frame\n",
    "    if not is_neut_standard:\n",
    "        df = df.query(\"serum == 'none'\")\n",
    "    df = (\n",
    "        df.assign(\n",
    "            sample_counts=lambda x: x.groupby(\"sample\")[\"count\"].transform(\"sum\"),\n",
    "            count_frac=lambda x: x[\"count\"] / x[\"sample_counts\"],\n",
    "            median_count_frac=lambda x: x.groupby(\"barcode\")[\"count_frac\"].transform(\n",
    "                \"median\"\n",
    "            ),\n",
    "            fold_change_from_median=lambda x: x[\"count_frac\"] / x[\"median_count_frac\"],\n",
    "        )\n",
    "        .drop(\n",
    "            columns=(\n",
    "                [\"sample\", \"serum_replicate\", \"sample_counts\", \"neut_standard\"]\n",
    "                + ([\"strain\"] if is_neut_standard else [\"dilution_factor\"])\n",
    "            ),\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # make chart\n",
    "    evenness_chart = (\n",
    "        alt.Chart(df)\n",
    "        .add_params(barcode_selection)\n",
    "        .encode(\n",
    "            alt.X(\n",
    "                \"count_frac\",\n",
    "                title=(\n",
    "                    \"barcode's fraction of neut standard counts\"\n",
    "                    if is_neut_standard\n",
    "                    else \"barcode's fraction of non-neut standard counts\"\n",
    "                ),\n",
    "                scale=alt.Scale(nice=False, padding=5),\n",
    "            ),\n",
    "            alt.Y(\n",
    "                \"sample_noplate\",\n",
    "                title=None,\n",
    "                sort=list(\n",
    "                    neut_standard_fracs.sort_values([\"serum\", \"dilution_factor\"])[\"sample_noplate\"]\n",
    "                ),\n",
    "            ),\n",
    "            alt.Fill(\"barcode\", legend=None),\n",
    "            strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "            size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3g\") if df[c].dtype == float\n",
    "                else c\n",
    "                for c in df.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_circle(fillOpacity=0.6, stroke=\"black\", strokeOpacity=1)\n",
    "        .properties(\n",
    "            height=alt.Step(10),\n",
    "            width=300,\n",
    "            title=(\n",
    "                f\"{plate} all samples, neut-standard barcodes\"\n",
    "                if is_neut_standard\n",
    "                else f\"{plate} no-serum samples, all barcodes\"\n",
    "            ),\n",
    "        )\n",
    "        .configure_axis(grid=False)\n",
    "        .configure_legend(titleLimit=1000)\n",
    "        .interactive()\n",
    "    )\n",
    "\n",
    "    if is_neut_standard:\n",
    "        evenness_chart = evenness_chart.add_params(serum_selection).transform_filter(\n",
    "            serum_selection\n",
    "        )\n",
    "        print(f\"\\n\\n{'=' * 89}\\nAnalyzing neut-standard barcodes from all samples\\n\")\n",
    "    else:\n",
    "        print(f\"\\n\\n{'=' * 89}\\nAnalyzing all barcodes from no-serum samples\\n\")\n",
    "\n",
    "    display(evenness_chart)\n",
    "\n",
    "    # make sure barcode fractions are reasonably consistent when they should be\n",
    "    excess_fold_change = df[\n",
    "        (df[\"fold_change_from_median\"] <= 1 / qc_thresholds[\"barcode_frac_consistency\"])\n",
    "        | (df[\"fold_change_from_median\"] >= qc_thresholds[\"barcode_frac_consistency\"])\n",
    "    ]\n",
    "    if len(excess_fold_change):\n",
    "        print(f\"\\nFollowing barcodes failed {qc_thresholds['barcode_frac_consistency']=}\")\n",
    "        display(excess_fold_change)\n",
    "        qc_failures.add(\"barcode_frac_consistency\")\n",
    "    else:\n",
    "        print(f\"\\nPassed {qc_thresholds['barcode_frac_consistency']=}\")\n",
    "\n",
    "    # make sure barcodes have sufficient fraction\n",
    "    if is_neut_standard:\n",
    "        insufficient_neut_standard_barcode_frac = df[\n",
    "            df[\"count_frac\"] < qc_thresholds[\"min_neut_standard_barcode_frac\"]\n",
    "        ]\n",
    "        if len(insufficient_neut_standard_barcode_frac):\n",
    "            print(\n",
    "                \"\\nFollowing barcodes fail \"\n",
    "                + f\"{qc_thresholds['min_neut_standard_barcode_frac']=}\"\n",
    "            )\n",
    "            display(insufficient_neut_standard_barcode_frac)\n",
    "            qc_failures.add(\"min_neut_standard_barcode_frac\")\n",
    "        else:\n",
    "            print(f\"\\nPassed {qc_thresholds['min_neut_standard_barcode_frac']=}\")\n",
    "    else:\n",
    "        insufficient_viral_barcode_frac = df[\n",
    "            df[\"count_frac\"] < qc_thresholds[\"min_viral_barcode_frac\"]\n",
    "        ]\n",
    "        if len(insufficient_viral_barcode_frac):\n",
    "            print(\n",
    "                f\"\\nFollowing barcodes fail {qc_thresholds['min_viral_barcode_frac']=}\"\n",
    "            )\n",
    "            display(insufficient_viral_barcode_frac)\n",
    "            qc_failures.add(\"min_viral_barcode_frac\")\n",
    "        else:\n",
    "            print(f\"\\nPassed {qc_thresholds['min_viral_barcode_frac']=}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743e4e41-aef9-4980-8b59-55d7b0a686fd",
   "metadata": {},
   "source": [
    "## Compute fraction infectivity\n",
    "\n",
    "The fraction infectivity for viral barcode $v_b$ in sample $s$ is computed as:\n",
    "$$\n",
    "F_{v_b,s} = \\frac{c_{v_b,s} / \\left(\\sum_{n_b} c_{n_b,s}\\right)}{{\\rm median}_{s_0}\\left[ c_{v_b,s_0} / \\left(\\sum_{n_b} c_{n_b,s_0}\\right)\\right]}\n",
    "$$\n",
    "where\n",
    " - $c_{v_b,s}$ is the counts of viral barcode $v_b$ in sample $s$.\n",
    " - $\\sum_{n_b} c_{n_b,s}$ is the sum of the counts for all neutralization standard barcodes $n_b$ for sample $s$.\n",
    " - $c_{v_b,s_0}$ is the counts of viral barcode $v_b$ in no-serum sample $s_0$.\n",
    " - $\\sum_{n_b} c_{n_b,s_0}$ is the sum of the counts for all neutralization standard barcodes $n_b$ for no-serum sample $s_0$.\n",
    " - ${\\rm median}_{s_0}\\left[ c_{v_b,s_0} / \\left(\\sum_{n_b} c_{n_b,s_0}\\right)\\right]$ is the median taken across all no-serum samples of the counts of viral barcode $v_b$ versus the total counts for all neutralization standard barcodes.\n",
    "\n",
    "First, compute the total neutralization-standard counts for each sample.\n",
    "Plot these, and make sure they meet the QC threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ef1afa4-5772-4596-a68d-b98e080ba2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "neut_standard_counts = (\n",
    "    counts\n",
    "    .query(\"neut_standard\")\n",
    "    .groupby(\n",
    "        [\"well\", \"serum\", \"sample_noplate\", \"dilution_factor\"], dropna=False, as_index=False,\n",
    "    )\n",
    "    .aggregate(neut_standard_count=pd.NamedAgg(\"count\", \"sum\"))\n",
    "    .assign(\n",
    "        passes_qc=lambda x: (\n",
    "            x[\"neut_standard_count\"] >= qc_thresholds[\"min_neut_standard_count\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "neut_standard_counts_chart = (\n",
    "    alt.Chart(neut_standard_counts)\n",
    "    .add_params(serum_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"neut_standard_count\",\n",
    "            title=\"counts from neutralization standard\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"sample_noplate\",\n",
    "            title=None,\n",
    "            sort=list(\n",
    "                neut_standard_counts.sort_values([\"serum\", \"dilution_factor\"])[\"sample_noplate\"]\n",
    "            ),\n",
    "        ),\n",
    "        alt.Color(\n",
    "            \"passes_qc\",\n",
    "            title=f\"at least {qc_thresholds['min_neut_standard_count']} counts\",\n",
    "            scale=alt.Scale(domain=[True, False]),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if neut_standard_counts[c].dtype == float\n",
    "            else c\n",
    "            for c in neut_standard_counts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Neutralization-standard counts for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(neut_standard_counts_chart)\n",
    "\n",
    "if (neut_standard_counts[\"passes_qc\"]).all():\n",
    "    print(f\"\\nAll samples pass {qc_thresholds['min_neut_standard_count']=}\")\n",
    "else:\n",
    "    print(f\"\\nSamples failing {qc_thresholds['min_neut_standard_count']=}\")\n",
    "    display(neut_standard_counts.query(\"not passes_qc\"))\n",
    "    qc_failures.add(\"min_neut_standard_count\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9baafb57-e601-46d4-bed4-4bd5eb61552b",
   "metadata": {},
   "source": [
    "Compute and plot the no-serum sample viral barcode counts and check if they pass the QC filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d060ad12-b159-49e0-8ff4-94d175da5bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_serum_counts = (\n",
    "    counts\n",
    "    .query(\"serum == 'none'\")\n",
    "    .query(\"not neut_standard\")\n",
    "    .merge(neut_standard_counts, validate=\"many_to_one\")\n",
    "    [[\"barcode\", \"strain\", \"well\", \"sample_noplate\", \"count\", \"neut_standard_count\"]]\n",
    "    .assign(\n",
    "        passes_qc=lambda x: (\n",
    "            x[\"count\"] >= qc_thresholds[\"min_no_serum_viral_barcode_count\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "# make chart\n",
    "no_serum_counts_chart = (\n",
    "    alt.Chart(no_serum_counts)\n",
    "    .add_params(barcode_selection)\n",
    "    .encode(\n",
    "        alt.X(\"count\", title=\"viral barcode count\", scale=alt.Scale(nice=False, padding=5)),\n",
    "        alt.Y(\"sample_noplate\", title=None),\n",
    "        alt.Fill(\"barcode\", legend=None),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "        size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "        tooltip=no_serum_counts.columns.tolist(),\n",
    "    )\n",
    "    .mark_circle(fillOpacity=0.6, stroke=\"black\", strokeOpacity=1)\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=300,\n",
    "        title=f\"{plate} viral barcode counts in no-serum samples\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    "    .interactive()\n",
    ")\n",
    "\n",
    "display(no_serum_counts_chart)\n",
    "\n",
    "# QC check\n",
    "if (no_serum_counts[\"passes_qc\"]).all():\n",
    "    print(f\"\\nAll samples pass {qc_thresholds['min_no_serum_viral_barcode_count']=}\")\n",
    "else:\n",
    "    print(f\"\\nSamples failing {qc_thresholds['min_no_serum_viral_barcode_count']=}\")\n",
    "    display(no_serum_counts.query(\"not passes_qc\"))\n",
    "    qc_failures.add(\"min_no_serum_viral_barcode_count\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdffe0e7-9864-4c4f-851d-5604106ff63a",
   "metadata": {},
   "source": [
    "Compute and plot the median ratio of viral barcode count to neut standard counts across no-serum samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b15a0c-9d54-48f4-86e3-f38b2015a5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "median_no_serum_ratio = (\n",
    "    no_serum_counts\n",
    "    .assign(ratio=lambda x: x[\"count\"] / x[\"neut_standard_count\"])\n",
    "    .groupby([\"barcode\", \"strain\"], as_index=False)\n",
    "    .aggregate(median_no_serum_ratio=pd.NamedAgg(\"ratio\", \"median\"))\n",
    ")\n",
    "\n",
    "strain_selection = alt.selection_point(fields=[\"strain\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "median_no_serum_ratio_chart = (\n",
    "    alt.Chart(median_no_serum_ratio)\n",
    "    .add_params(strain_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"median_no_serum_ratio\",\n",
    "            title=\"median ratio of counts\",\n",
    "            scale=alt.Scale(nice=False, padding=5),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"barcode\",\n",
    "            sort=alt.SortField(\"median_no_serum_ratio\", order=\"descending\"),\n",
    "            axis=alt.Axis(labelFontSize=5),\n",
    "        ),\n",
    "        color=alt.condition(strain_selection, alt.value(\"orange\"), alt.value(\"gray\")),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if median_no_serum_ratio[c].dtype == float\n",
    "            else c\n",
    "            for c in median_no_serum_ratio.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar(height={\"band\": 0.85})\n",
    "    .properties(\n",
    "        height=alt.Step(5),\n",
    "        width=250,\n",
    "        title=f\"{plate} no-serum median ratio viral barcode to neut-standard barcode\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(median_no_serum_ratio_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb39004a-3684-4501-afd0-da259c85b131",
   "metadata": {},
   "source": [
    "Compute the actual fraction infectivities, QC check if any are null (from zero counts), and also plot and check if any exceed the `max_frac_infectivity`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d787193c-3c50-427d-a0ec-165b446d8e73",
   "metadata": {},
   "outputs": [],
   "source": [
    "frac_infectivity = (\n",
    "    counts\n",
    "    .query(\"not neut_standard\")\n",
    "    .query(\"serum != 'none'\")\n",
    "    .merge(median_no_serum_ratio, validate=\"many_to_one\")\n",
    "    .merge(\n",
    "        neut_standard_counts.drop(columns=\"passes_qc\"),\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .assign(\n",
    "        frac_infectivity=lambda x: (\n",
    "            (x[\"count\"] / x[\"neut_standard_count\"]) / x[\"median_no_serum_ratio\"]\n",
    "        ),\n",
    "        passes_qc=lambda x: x[\"frac_infectivity\"] <= qc_thresholds[\"max_frac_infectivity\"],\n",
    "    )\n",
    "    [\n",
    "        [\n",
    "            \"barcode\",\n",
    "            \"strain\",\n",
    "            \"serum\",\n",
    "            \"serum_replicate\",\n",
    "            \"plate_replicate\",\n",
    "            \"dilution_factor\",\n",
    "            \"frac_infectivity\",\n",
    "            \"sample_noplate\",\n",
    "            \"well\",\n",
    "            \"passes_qc\",\n",
    "        ]\n",
    "    ]\n",
    ")\n",
    "\n",
    "assert (\n",
    "    len(frac_infectivity.groupby([\"barcode\", \"serum\", \"plate_replicate\", \"dilution_factor\"]))\n",
    "    == len(frac_infectivity)\n",
    ")\n",
    "assert frac_infectivity[\"dilution_factor\"].notnull().all()\n",
    "\n",
    "frac_infectivity_chart = (\n",
    "    alt.Chart(frac_infectivity)\n",
    "    .add_params(serum_selection, barcode_selection)\n",
    "    .transform_filter(serum_selection)\n",
    "    .encode(\n",
    "        alt.X(\n",
    "            \"frac_infectivity\",\n",
    "            title=\"fraction infectivity\",\n",
    "            scale=alt.Scale(nice=False, padding=3),\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"sample_noplate\",\n",
    "            title=None,\n",
    "            sort=list(\n",
    "                neut_standard_counts.sort_values([\"serum\", \"dilution_factor\"])[\"sample_noplate\"]\n",
    "            ),\n",
    "        ),\n",
    "        strokeWidth=alt.condition(barcode_selection, alt.value(2), alt.value(0)),\n",
    "        size=alt.condition(barcode_selection, alt.value(60), alt.value(35)),\n",
    "        color=alt.Color(\n",
    "            \"passes_qc\",\n",
    "            title=f\"frac_infectivity <= {qc_thresholds['max_frac_infectivity']}\",\n",
    "            scale=alt.Scale(domain=[True, False]),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if frac_infectivity[c].dtype == float\n",
    "            else c\n",
    "            for c in frac_infectivity.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_circle(stroke=\"black\", strokeOpacity=1)\n",
    "    .properties(\n",
    "        height=alt.Step(10),\n",
    "        width=250,\n",
    "        title=f\"Fraction infectivities for {plate}\",\n",
    "    )\n",
    "    .configure_axis(grid=False)\n",
    "    .configure_legend(titleLimit=1000)\n",
    ")\n",
    "\n",
    "display(frac_infectivity_chart)\n",
    "\n",
    "if not frac_infectivity[\"passes_qc\"].all():\n",
    "    print(f\"\\nSome barcode-samples fail {qc_thresholds['max_frac_infectivity']=}\")\n",
    "    display(frac_infectivity.query(\"not passes_qc\"))\n",
    "    qc_failures.add(\"max_frac_infectivity\")\n",
    "else:\n",
    "    print(\"\\nAll barcode-samples pass {qc_thresholds['max_frac_infectivity']=}\")\n",
    "\n",
    "if frac_infectivity[\"frac_infectivity\"].isnull().any():\n",
    "    print(\"\\nSome barcodes have undefined fraction infectivity due to zero counts:\")\n",
    "    display(frac_infectivity.query(\"frac_infectivity.isnull()\"))\n",
    "    qc_failures.add(\"null_frac_infectivity\")\n",
    "else:\n",
    "    print(\"\\nNo undefined fraction infectivities\")\n",
    "\n",
    "frac_infectivity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "359efaf0-4be9-44bc-839d-511c8afc70db",
   "metadata": {},
   "source": [
    "Write fraction infectivities to file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70e5e6cc-f729-4509-8cb3-fabeefe38033",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"\\nWriting fraction infectivities to {frac_infectivity_csv}\")\n",
    "(\n",
    "    frac_infectivity\n",
    "    [[\"barcode\", \"strain\", \"serum\", \"plate_replicate\", \"dilution_factor\", \"frac_infectivity\"]]\n",
    "    .sort_values([\"serum\", \"plate_replicate\", \"dilution_factor\", \"barcode\"])\n",
    "    .to_csv(frac_infectivity_csv, index=False, float_format=\"%.5g\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdb4eea-9822-4f27-89e6-6f11033bcb7f",
   "metadata": {},
   "source": [
    "Make sure we have enough dilutions with non-null fraction infectivities for each serum-replicate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e8998c6-b2de-43b8-9091-b65423257bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_dilutions = (\n",
    "    frac_infectivity\n",
    "    .query(\"frac_infectivity.notnull()\")\n",
    "    .groupby(\"serum_replicate\")\n",
    "    .aggregate(n_dilutions=pd.NamedAgg(\"dilution_factor\", \"nunique\"))\n",
    "    .assign(\n",
    "        fails_qc=lambda x: (\n",
    "            x[\"n_dilutions\"] <= qc_thresholds[\"min_dilutions_per_serum_replicate\"]\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "if n_dilutions[\"fails_qc\"].any():\n",
    "    print(f\"Failing {qc_thresholds['min_dilutions_per_serum_replicate']=}:\")\n",
    "    display(n_dilutions.query(\"fails_qc\"))\n",
    "    qc_failures.add(\"min_dilutions_per_serum_replicate\")\n",
    "else:\n",
    "    print(f\"Passed {qc_thresholds['min_dilutions_per_serum_replicate']=}:\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "944dab61-fd42-42f2-914f-ca9e77440911",
   "metadata": {},
   "source": [
    "Summarize all QC failures and write to file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2adb888-dec4-4b5d-9bed-4be7c21c5f27",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_failures = \"\\n\".join(sorted(qc_failures))\n",
    "\n",
    "if qc_failures:\n",
    "    print(f\"Encountered the following QC failures:\\n{qc_failures}\")\n",
    "else:\n",
    "    print(\"No QC failures\")\n",
    "\n",
    "print(f\"\\nLogging QC failures to {qc_failures_file}\")\n",
    "with open(qc_failures_file, \"w\") as f:\n",
    "    f.write(qc_failures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b973c7fa-138f-4174-a8d1-cbffffaa7a8c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
